---
output: rtf_document
editor_options:
  chunk_output_type: console
---

---
bibliography: 
- `r here::here("ms/references.yaml")`
---

```{r manual, include = FALSE, evaluate = FALSE}
# Run these lines if rendering outside of targets plan
library(targets)
library(tarchetypes)
source(here::here("R/packages.R"))
source(here::here("R/functions.R"))
source(here::here("R/ms_functions.R"))
```

```{r setup, include = FALSE, cache = FALSE}
# TO CONVERT TO PLAIN TEXT:
# after rendering to rtf,
# 1. Copy the contents of the rtf file, paste to txt.
# 2. Replace all single line breaks with double line breaks
# (in Atom search-replace with grep: \n for \n\n)
# 3. Hard-wrap all text at 80 characters (in Atom with autoflow: cmd-alt-q)
# (optional: add name after first line)
# 4. Save as "README.txt"

knitr::opts_chunk$set(
  echo = FALSE, message = FALSE,
  warning = FALSE, results = "hide", cache = FALSE)

# Load objects from targets workflow (all paths to data files)
tar_load(
  c(
    "ms_functions",
    "ebihara_2019_zip_file",
    "ftol_zip_file",
    "ja_climate_data_file",
    "japan_map_file",
    "deer_range_file",
    "japan_ferns_comm_full_file",
    "inext_res_file",
    "occ_point_data_summary_file",
    "plastome_calibration_dates",
    "redundancy_by_res_file",
    "japan_ferns_shape_full_file",
    "raw_trait_data_file",
    "japan_map_points_file",
    "protected_areas_file",
    "calibration_dates_file",
    "japan_ferns_biodiv_dryad_file",
    "comm_ferns_dryad_file",
    "shape_ferns_dryad_file",
    "fern_traits_dryad_file",
    "japan_fern_phylogram_dryad_file",
    "japan_fern_tree_dryad_file",
    "japan_fern_tree_uncollapsed_dryad_file",
    "refs_yaml"
    )
  )
```

This README.txt file was generated on `r format(Sys.time(), '%d %B, %Y')` by Joel Nitta

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- GENERAL INFORMATION \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Title of Dataset: Data from: Spatial phylogenetics of Japanese ferns: Patterns,
processes, and implications for conservation

Author Information

Principal Investigator: Joel H. Nitta

Department of Biological Sciences, Graduate School of Science, The University of
Tokyo, 2-11-16 Yayoi, Bunkyo-ku, Tokyo 113-0032, Japan joelnitta@gmail.com

Associate or Co-investigators: Brent D. Mishler, Wataru Iwasaki, Atsushi Ebihara

Geographic location of data collection: Japan

Information about funding sources or sponsorship that supported the collection
of the data: Funding provided in part by the Japan Society for the Promotion of
Science (Kakenhi) Grant Number 16H06279

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

SHARING/ACCESS INFORMATION

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Licenses/restrictions placed on the data, or limitations of reuse: CC0 1.0
Universal (CC0 1.0)

Recommended citation for the data: Nitta JH, Mishler BD, Iwasaki W, Ebihara A
(2021) Data from: Spatial phylogenetics of Japanese ferns: Patterns, processes,
and implications for conservation FIXME: add DOI when available

Citation for and links to publications that cite or use the data: Nitta JH,
Mishler BD, Iwasaki W, Ebihara A (2021) Spatial phylogenetics of Japanese ferns:
Patterns, processes, and implications for conservation FIXME: add journal when published

Code for analyzing the data is available on github:
https://github.com/joelnitta/japan_ferns_spatial_phy

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

DATA & FILE OVERVIEW

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

File list (filenames, directory structure (for zipped files) and brief
description of all data files):

- _targets.tar.gz: Tarball (compressed folder) including all workflow results produced by R targets package
- japan_climate.gpkg: Climate data in Japan downloaded from WorldClim database
- japan_deer_range.gpkg: Distribution maps of Japanese deer (Cervus nippon) in Japan
- japan_ferns_comm_full.csv: Community matrix (species x sites matrix) of native, non-hybrid ferns in Japan, full (unfiltered) dataset
- japan_ferns_inext_results.csv: Results of collection curve analysis on fern specimens in Japan
- japan_ferns_occ_summary.csv: Summary of specimen data of native, non-hybrid ferns in Japan
- japan_ferns_redundancy_by_res: Taxonomic richness, abundance, and redunancy at different spatial resolutions for native, non-hybrid ferns of Japan
- japan_ferns_shape_full.gpkg: Location of grid-cells (sites) for native, non-hybrid ferns in Japan, full (unfiltered) dataset
- japan_ferns_traits_lucid.csv: Morphological trait data of native, non-hybrid ferns in Japan
- japan_map_points.csv: Latitude and longitude of points used to produce a labeled map of Japan
- japan_map.gpkg: Map of Japan
- japan_protected_areas.gpkg: Protected areas in Japan
- testo_sundue_2016_calibrations.csv: Fossil calibration points used for phylogenetic dating analysis
- results.zip: Selected analysis results. Files include:
  - japan_ferns_biodiv.csv: Biodiversity statistics of native, non-hybrid ferns in Japan
  - japan_ferns_comm.csv: Community matrix (species x sites matrix) of native, non-hybrid ferns in Japan used for biodiversity analysis
  - japan_ferns_shape.gpkg: Location of grid-cells (sites) for native, non-hybrid ferns in Japan used for biodiversity analysis
  - japan_ferns_traits.csv: Trait matrix of native, non-hybrid ferns in Japan used for functional biodiversity analysis
  - japan_ferns_tree_dated.tre: Maximim-likelihood, ultrametric (dated) phylogeny of native, non-hybrid ferns in Japan with clades consisting of identical OTUs collapsed to polytomies
  - japan_ferns_tree_uncollapsed.tre: Maximim-likelihood, ultrametric (dated) phylogeny of native, non-hybrid ferns in Japan with clades consisting of identical OTUs not collapsed
  - japan_ferns_tree.tre: Maximim-likelihood phylogeny of native, non-hybrid ferns in Japan

Additional data used in analysis not included in current data package: 

- doi_10.5061_dryad.4362p32__v4.zip: Data including DNA sequences, taxonomy, and 
breeding system data for native Japanese ferns [@Ebihara2019b] downloaded from 
https://datadryad.org/stash/dataset/doi:10.5061/dryad.4362p32. 
MD5 checksum `r tools::md5sum(here::here(ebihara_2019_zip_file))`
- ftol_data_release_v0.0.1.zip: Fern Tree of Life project global DNA alignment 
and phylogenetic tree (Nitta et al., in prep.) downloaded from 
https://figshare.com/s/87b8ec102303bf9f3212.
MD5 checksum `r tools::md5sum(here::here(ftol_zip_file))`
FIXME: replace with DOI when available

Checksums are 32-byte MD5 hashes generated with tools::md5sum() in R.

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

METHODOLOGICAL INFORMATION

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r summary-stats}
# Number of fossils used for dating
n_fossil_points <- plastome_calibration_dates %>%
  filter(age_type != "fixed") %>%
  nrow()
```

Description of methods used for collection/generation of data:

A list of native, non-hybrid fern specimens housed at the herbarium of the
Museum of Science and Nature, Japan was converted to a community data matrix at
four grain sizes (grid-cells spanning Japan, each 0.1, 0.2, 0.3, or 0.4 degrees
per side). The 0.2 degree grain size was selected for further analysis based on 
redundancy (ratio of number of specimens to number of taxa per cell).

All taxon names are based on the Green List (http://www.rdplants.org/gl/;
English version available at
https://datadryad.org/stash/dataset/doi:10.5061/dryad.4362p32).

Traits were measured on each species as described in @Ebihara2019b.

Phylogenetic analysis was conducted with maximum likelihood in IQ-TREE v1.6.12
[@Nguyen2015] by combining plastid *rbcL* sequences of each taxon with a
globally sampled data matrix (Nitta et al, in prep). Next, dating analysis was
carried out using treePL v1.0 [@Smith2012] with `r n_fossil_points` fossil
calibration points after @Testo2016a. The dated phylogeny was then trimmed to
include Japanese taxa only.

The community matrix, traits, and phylogeny were used to analyze spatial
patterns of phylogenetic diversity and endemism.

Data files were generated from raw data (not included here) using scripts
available at https://github.com/joelnitta/japan_ferns_spatial_phy, in particular
https://github.com/joelnitta/japan_ferns_spatial_phy/blob/main/R/process_raw_data.R.

For full methods, see Nitta JH, Mishler BD, Iwasaki W, Ebihara A (2021) Spatial
phylogenetics of Japanese ferns: Patterns, processes, and implications for
conservation FIXME: Add journal when published

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

DATA-SPECIFIC INFORMATION

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

_targets.tar.gz: Tarball (compressed folder) including all workflow results
produced by R targets package. This is provided to enable inspection of workflow
steps without running the entire workflow from the beginning. To use it, unpack
the tar achive with the command "tar -xzf _targets.tar.gz". Then, in R, the
"tar_load()" function in the R package "targets" can be used to load any
workflow step (target) defined in _targets.R
(https://github.com/joelnitta/japan_ferns_spatial_phy/blob/main/_targets.R). For
more information on the structure of the _targets folder and how to use it, see
https://github.com/ropensci/targets.

MD5 checksum: FIXME (add manually, since this can't be calculated from inside the targets workflow)

Corresponding commit in repo (https://github.com/joelnitta/japan_ferns_spatial_phy): FIXME add manually

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r deer-range}
deer_range <- sf::st_read(here::here(deer_range_file))
# Check that no data are missing
assert(deer_range, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(deer_range_file)`: Distribution maps of Japanese deer (Cervus
nippon) in Japan. Raw data were downloaded from the Japan Ministry of the
Environment
(https://www.biodic.go.jp/biodiversity/activity/policy/map/map14/index.html).
The data include range maps based on three types of data: range of Japanese deer
surveyed in 1978, range surveyed in 2003, and estimated range inferred from a
model including snow cover and forest type based on the 2003 survey data. Only
estimated range with < 0.10 movement cost was included.

Number of variables: `r ncol(deer_range)`

Number of cases/rows: `r nrow(deer_range)`

Coordinate reference system: JGD2000

Variable list:

```{r results = "asis"}
  deer_range %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "range", "Source of range data. '1978' = 1978 survey, '2003' = 2003 survey, 'estimated' = range estimated from model based on 2003 survey data",
      "geom", "Vector describing shape and position of range polygon"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(deer_range_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r ja-climate-data}
ja_climate_data <- sf::st_read(here::here(ja_climate_data_file))
# Check that no data are missing
assert(ja_climate_data, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(ja_climate_data_file)`: Climate data in Japan downloaded from
WorldClim database (https://worldclim.org/data/v1.4/formats.html). Raw data were
downloaded at the 2.5 minute scale. The intersection of the raw climate data
with 0.2 degree grid-cells (japan_ferns_shape_full.gpkg) were identified,
then average value for each climatic variable was calculated for each grid-cell.

Number of variables: `r ncol(ja_climate_data)`

Number of cases/rows: `r nrow(ja_climate_data)`

Coordinate reference system: JGD2000

Variable list:

```{r ja-climate-data-cols, results = "asis"}
  ja_climate_data %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "temp", "Mean annual temperature, in units of 10 * degrees celsius (BIO1)",
      "temp_season", "Temperature seasonality (standard deviation Ã—100) (BIO4)",
      "precip", "Annual precipitation (mm) (BIO12)",
      "precip_season", "Precipitation seasonlaity (coefficient of variation) (BIO15)",
      "geom", "Vector describing shape and position of grid-cell"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(ja_climate_data_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-ferns-comm-full}
japan_ferns_comm_full <- read_csv(here::here(japan_ferns_comm_full_file))
# Check that no data are missing
assert(japan_ferns_comm_full, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(japan_ferns_comm_full_file)`: Community matrix (species x sites
matrix) of native, non-hybrid ferns in Japan. Sites correspond to 0.2 degree x
0.2 degree grid-cells covering Japan. For location of grid-cells, see
japan_ferns_shape_full.gpkg. Full dataset including all grid-cells (none
excluded due to low redundancy).

Number of variables: `r ncol(japan_ferns_comm_full)`

Number of cases/rows: `r nrow(japan_ferns_comm_full)`

Variable list:

- grids: Name of grid-cell.
- Other columns: Each column is named for a Japanese fern taxon. 
  Values indicate number of specimens that were observed in each grid-cell for that taxon.

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_ferns_comm_full_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-ferns-inext-results}
inext_res <- read_csv(here::here(inext_res_file))
# Check that no data are missing
assert(inext_res, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(inext_res_file)`: Output from running iNEXT [@Hsieh2016] on
specimen data of native, non-hybrid ferns in Japan. iNEXT estimates diversity
(here, species richness) and sampling completeness based on interpolation and
extrapolation from the original data.

Number of variables: `r ncol(inext_res)`

Number of cases/rows: `r nrow(inext_res)`

Variable list:

```{r japan-ferns-inext-results-cols, results = "asis"}
  inext_res %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "m", "Sample size (number of individuals)",
      "method", "Method used to estimate diveristy",
      "order", "Diversity order of Hill number (here 0, equivalent to species richness)",
      "qD", "Estimated diversity",
      "qD.LCL", "Lower 95% confidence limit on estimated diversity",
      "qD.UCL", "Upper 95% confidence limit on estimated diversity",
      "SC", "Estimated sampling completeness",
      "SC.LCL", "Lower 95% confidence limit on sampling completeness",
      "SC.UCL", "Upper 95% confidence limit on sampling completeness")
  )
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(inext_res_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r occ-point-data-summary}
occ_point_data_summary <- read_csv(here::here(occ_point_data_summary_file))
# Check that no data are missing
assert(occ_point_data_summary, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(occ_point_data_summary_file)`: Summary of raw specimen data for
native, non-hybrid ferns of Japan. Raw specimen data consists of one row per
specimen, with species name, latitude, longitude, and collection date (raw
specimen data not included here to protect endangered species). Raw data were
filtered to remove duplicate specimens and specimens that occurred outside of
the second-degree mesh (approximately 10 km sq grid-cell) map of Japan
(http://www.biodic.go.jp/trialSystem/EN/shpddl.html)

Number of variables: `r ncol(occ_point_data_summary)`

Number of cases/rows: `r nrow(occ_point_data_summary)`

Variable list:

```{r occ-point-data-summary-cols, results = "asis"}
  occ_point_data_summary %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "dataset", "Dataset name. 'occ_point_data_ferns_unfiltered' indicates data includes all specimens, 'occ_point_data_ferns' indicates data only includes specimens after filtering",
      "variable", "Variable. 'n_taxa' indicates number of taxa, 'n_specimens' indicates number of specimens",
      "value", "Value corresponding to each variable")
  )
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(occ_point_data_summary_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r redundancy-by-res}
redundancy_by_res <- read_csv(here::here(redundancy_by_res_file))
# Check that no data are missing
assert(redundancy_by_res, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(redundancy_by_res_file)`: Results of testing binning raw
specimen data of native, non-hybrid ferns of Japan into grid-cells at different
resolutions. Sampling redundancy is a metric of sampling completeness,
calculated as 1 - (richness/abundance).

Number of variables: `r ncol(redundancy_by_res)`

Number of cases/rows: `r nrow(redundancy_by_res)`

Variable list:

```{r redundancy-by-res-cols, results = "asis"}
  redundancy_by_res %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "res", "Degree of resolution (grid-cell side length), from 0.1 to 0.4 degrees",
      "grids", "Name of grid-cell",
      "abundance", "Number of specimens occurring in grid-cell",
      "richness", "Number of taxa occurring in grid-cell",
      "redundancy", "Redundancy (measure of sampling completeness)"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(redundancy_by_res_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-ferns-shape-full}
japan_ferns_shape_full <- st_read(here::here(japan_ferns_shape_full_file))
# Check that no data are missing
assert(japan_ferns_shape_full, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(japan_ferns_shape_full_file)`: Location of grid-cells (sites)
for native, non-hybrid ferns in Japan in GeoPackage format. Compiled from raw
specimen data by binning specimens into 0.2 degree grid-cells across Japan. Full
dataset including all grid-cells (none excluded due to low redundancy).

Number of variables: `r ncol(japan_ferns_shape_full)`

Number of cases/rows: `r nrow(japan_ferns_shape_full)`

Coordinate reference system: JGD2000

Variable list:

```{r japan-ferns-shape-full-cols, results = "asis"}
  japan_ferns_shape_full %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "grids", "Name of grid-cell",
      "abundance", "Number of specimens occurring in grid-cell",
      "richness", "Number of taxa occurring in grid-cell",
      "redundancy", "Redundancy (measure of sampling completeness)",
      "geom", "Vector describing shape and position of grid-cell"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_ferns_shape_full_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-ferns-traits-lucid}
japan_ferns_traits_lucid <- read_csv(here::here(raw_trait_data_file))
# Check that no data are missing
assert(japan_ferns_traits_lucid, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(raw_trait_data_file)`: Morphological trait data of native,
non-hybrid ferns in Japan, originally formatted for Lucid dichotomous key
software (https://www.lucidcentral.org/). For more information on codes used for
scoring of traits, see https://help.lucidcentral.org/lucid/scoring-the-key/

Number of variables: `r ncol(japan_ferns_traits_lucid)`

Number of cases/rows: `r nrow(japan_ferns_traits_lucid)`

Variable list: 'taxon' indicates taxon name. Other columns correspond to traits.
If the column name includes 'number' it is numeric, otherwise categorical.

Categorical trait codes as follows:

- 0 = absent
- 1 = common
- 2 = rare
- 3 = uncertain
- 4 = common and misinterpreted
- 5 = rare and misinterpreted
- 6 = not scoped

Numeric trait data formatted as a series of numbers separated by colons (':'),
unless missing (0 or 3). For example, 1:9.4:12:14:16. The starting '1:' does not
mean anything. The next series of numbers correspond to the outside minimum,
normal minimum, normal maximum and outside maximum. For example, a plant leaf
may normally be in the range of 4-5 cm, but sometimes as low as 2 cm or as high
as 6 cm. This would be written 2:4:5:6.

Missing data codes: '0' indicates trait is missing; '3' indicates uncertain.

MD5 checksum: `r tools::md5sum(here::here(raw_trait_data_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-map-points}
japan_map_points <- read_csv(here::here(japan_map_points_file))
```

`r fs::path_file(japan_map_points_file)`: Latitude and longitude of points used
to produce a labeled map of Japan. Some points were originally generated using
script at https://github.com/joelnitta/japan_ferns_spatial_phy/blob/main/R/geocode.R

Number of variables: `r ncol(japan_map_points)`

Number of cases/rows: `r nrow(japan_map_points)`

Variable list:

```{r japan-map-points-cols, results = "asis"}
  japan_map_points %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "query", "Name of point of interest",
      "lat", "Latitude",
      "lon", "Longitude",
      "lat_end", "Ending latitude (only for features that are lines, not points)",
      "lon_end", "Ending longitude (only for features that are lines, not points)"
  ))
```

Missing data codes: No data entered between commas in CSV.

MD5 checksum: `r tools::md5sum(here::here(japan_map_points_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-map}
japan_map <- sf::st_read(here::here(japan_map_file))
# Check that no data are missing
assert(japan_map, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(japan_map_file)`: Map of land area of Japan. Raw data downloaded from
Geospatial Information Authority of Japan under the Creative Commons Attribution
License v4.0 (https://www.gsi.go.jp/kankyochiri/gm_japan_e.html). All polygons
were combined into a single polygon and CRS set to JGD2000 using code available at
https://github.com/joelnitta/japan_ferns_spatial_phy/blob/main/R/process_raw_data.R.

Number of variables: `r ncol(japan_map)`

Number of cases/rows: `r nrow(japan_map)`

Variable list:

```{r japan-map-cols, results = "asis"}
  japan_map %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "geom", "Vector describing shape and position of land area polygon"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_map_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r protected-areas}
protected_areas <- st_read(here::here(protected_areas_file))
# Check that no data are missing
assert(protected_areas, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(protected_areas_file)`: Spatial of protected areas in Japan
in GeoPackage format. Original spatial data downloaded from the Japan Ministry
of the Environment
(https://www.biodic.go.jp/biodiversity/activity/policy/map/map17/index.html) and
Ministry of Land, Infrastructure, Transport and Tourism
(https://nlftp.mlit.go.jp/ksj/gml/datalist/KsjTmplt-A45.html). Categorization of
protected areas generally followed @Kusumoto2017: either "high" (no human
activities allowed at all) or "medium" status (some economic activities allowed
by permit); other areas not meeting these criteria or not including plants were
not considered. Any overlapping areas within the same protection level were
combined, and any areas overlapping between "medium" and "high" were considered
only "high". Code used to categorize areas available at
https://github.com/joelnitta/japan_ferns_spatial_phy/blob/main/R/process_raw_data.R

Number of variables: `r ncol(protected_areas)`

Number of cases/rows: `r nrow(protected_areas)`

Coordinate reference system: JGD2000

Variable list:

```{r protected-areas-cols, results = "asis"}
  protected_areas %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "status", "Protection status ('high' or 'medium')",
      "geom", "Vector describing shape and position of protected area polygon"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(protected_areas_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r plastome-calibration-dates}
plastome_calibration_dates <- read_csv(here::here(calibration_dates_file))
```

`r fs::path_file(calibration_dates_file)`: Fossil calibration points used for
phylogenetic dating analysis with treePL [@Smith2012], based on @Testo2016a.
MRCA = "most recent common ancestor".

Number of variables: `r ncol(plastome_calibration_dates)`

Number of cases/rows: `r nrow(plastome_calibration_dates)`

Variable list:

```{r plastome-calibration-dates-cols, results = "asis"}
  plastome_calibration_dates %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "Clade", "Name of clade",
      "Stem_Crown", "Type of node. 'crown' (node corresponding to MRCA of extant taxa only) or 'stem' (node corresponding to split between MRCA of extant taxa and its sister)",
      "Fossil", "Name of fossil",
      "Age", "Age of fossil",
      "Age_type", "Type of age. 'min' for minimum age, or 'fixed' for fixed age",
      "Citation", "Citation for the fossil",
      "taxon_1", "Name of first taxon that is used to identify MRCA",
      "taxon_2", "Name of second taxon that is used to identify MRCA",
      "note", "Notes, in particular explaining any differences with Testo and Sundue (2016)"
  ))
```

Missing data codes: No data entered between commas in CSV.

MD5 checksum: `r tools::md5sum(here::here(calibration_dates_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-ferns-biodiv-dryad}
japan_ferns_biodiv_dryad <- read_csv(here::here(japan_ferns_biodiv_dryad_file))
```

`r fs::path_file(japan_ferns_biodiv_dryad_file)` (contained in "results.zip"):
Biodiversity statistics of native, non-hybrid ferns and environmental variables
in Japan. Biodiversity metrics calculated as described in Nitta et al. 2021. 
Climatic (temperature and preciptation) variables
calculated as described for japan_climate.gpkg. Includes one row with missing
environmental data and one outlier for % apomixis that were removed prior to
spatial modeling analysis in Nitta et al. 2021.

Number of variables: `r ncol(japan_ferns_biodiv_dryad)`

Number of cases/rows: `r nrow(japan_ferns_biodiv_dryad)`

Variable list:

```{r japan-ferns-biodiv-dryad-cols, results = "asis"}
  japan_ferns_biodiv_dryad %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "grids", "Name of grid-cell",
      "lat", "Latitude of grid-cell centroid",
      "long", "Longitude of grid-cell centroid",
      "abundance", "Number of specimens occurring in grid-cell",
      "richness", "Number of taxa occurring in grid-cell",
      "fd_obs_z", "Z-score (standard effect size) of observed functional diversity (FD)",
      "pd_obs_z", "Z-score of observed phylogenetic diversity (PD)",
      "pe_obs_z", "Z-score of observed phylogenetic endemisim (PE)",
      "rfd_obs_z", "Z-score of observed relative FD",
      "rpd_obs_z", "Z-score of observed relative PD",
      "richness_obs_p_upper", "Richness percentile",
      "fd_obs_p_upper", "FD percentile",
      "pd_obs_p_upper", "PD percentile",
      "pe_obs_p_upper", "PE percentile",
      "pd_signif", "Significance of PD compared to null distribution, two-sided test at alpha 0.05",
      "rpd_signif", "Significance of RPD compared to null distribution, two-sided test at alpha 0.05",
      "fd_signif", "Significance of FD compared to null distribution, two-sided test at alpha 0.05",
      "rfd_signif", "Significance of RFD compared to null distribution, two-sided test at alpha 0.05",
      "pe_signif", "Significance of PE compared to null distribution, two-sided test at alpha 0.05",
      "taxonomic_cluster", "Membership in cluster based on taxonomic distances",
      "phylo_cluster", "Membership in cluster based on phylogenetic distances",
      "endem_type", "Endemism type determined by CANAPE",
      "lat_area", "Area of 1-degree latitudinal band around grid-cell",
      "temp", "Annual mean temperature, in units of degrees celsius * 10",
      "temp_season", "Temperature seasonality",
      "precip", "Annual precipitation, measured in mm",
      "precip_season", "Precipitation seasonality",
      "percent_apo", "Percent apomictic taxa"
    ))
```

Missing data codes: No data entered between commas in CSV.

MD5 checksum: `r tools::md5sum(here::here(japan_ferns_biodiv_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r comm-ferns-dryad-file}
comm_ferns_dryad <- read_csv(here::here(comm_ferns_dryad_file))
# Check that no data are missing
assert(comm_ferns_dryad, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(comm_ferns_dryad_file)` (contained in "results.zip"): Community
matrix (species x sites matrix) of native, non-hybrid ferns in Japan used for
biodiversity analysis. Same as japan_ferns_comm_full.csv, but grid-cells with
redundancy <0.1 (indicating under-sampling) excluded. 

Number of variables: `r ncol(comm_ferns_dryad)`

Number of cases/rows: `r nrow(comm_ferns_dryad)`

Variable list:

- grids: Name of grid-cell.
- Other columns: Each column is named for a Japanese fern taxon. 
  Values indicate number of specimens that were observed in each grid-cell for that taxon.

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(comm_ferns_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r shape-ferns-dryad-file}
shape_ferns_dryad <- sf::st_read(here::here(shape_ferns_dryad_file))
# Check that no data are missing
assert(shape_ferns_dryad, not_na, everything(), success_fun = success_logical)
```

`r fs::path_file(shape_ferns_dryad_file)` (contained in "results.zip"): Location
of grid-cells (sites) for native, non-hybrid ferns in Japan used for
biodiversity analysis. Same as japan_ferns_shape_full.gpkg, but grid-cells with
redundancy <0.1 (indicating under-sampling) excluded.

Number of variables: `r ncol(shape_ferns_dryad)`

Number of cases/rows: `r nrow(shape_ferns_dryad)`

Coordinate reference system: JGD2000

Variable list:

```{r shape-ferns-dryad-file-cols, results = "asis"}
  shape_ferns_dryad %>%
  make_data_desc(
    tribble(
      ~col, ~desc,
      "grids", "Name of grid-cell",
      "abundance", "Number of specimens occurring in grid-cell",
      "richness", "Number of taxa occurring in grid-cell",
      "redundancy", "Redundancy (measure of sampling completeness)",
      "geom", "Vector describing shape and position of grid-cell"
  ))
```

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(shape_ferns_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r fern-traits-dryad-file}
fern_traits_dryad <- read_csv(here::here(fern_traits_dryad_file))
```

`r fs::path_file(fern_traits_dryad_file)` (contained in "results.zip"): Trait
matrix of native, non-hybrid ferns in Japan used for functional biodiversity
analysis.

Number of variables: `r ncol(fern_traits_dryad)`

Number of cases/rows: `r nrow(fern_traits_dryad)`

Variable list: 'taxon' indicates taxon name. Other columns correspond to traits.
'frond_width', 'stipe_length', 'number_pinna_pairs' are numeric; others are
binary (0 or 1)

Missing data codes: Missing data coded as 'NA'.

MD5 checksum: `r tools::md5sum(here::here(fern_traits_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-fern-phylogram-dryad-file}
japan_fern_phylogram_dryad <- ape::read.tree(here::here(japan_fern_phylogram_dryad_file))
```

`r fs::path_file(japan_fern_phylogram_dryad_file)` (contained in
"results.zip"): Maximim-likelihood phylogeny of native, non-hybrid ferns in
Japan in newick format inferred with maximum likelihood IQ-TREE v1.6.12
[@Nguyen2015]. Values at nodes are bootstrap support values calculated using
1000 replicates of ultrafast bootstrap [@Nguyen2015].

Number of tips: `r ape::Ntip(japan_fern_phylogram_dryad)`

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_fern_phylogram_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-fern-dryad-uncollapsed-file}
japan_fern_tree_uncollapsed_dryad <- ape::read.tree(here::here(japan_fern_tree_uncollapsed_dryad_file))
```

`r fs::path_file(japan_fern_tree_uncollapsed_dryad_file)` (contained in "results.zip"):
Maximim-likelihood, ultrametric (dated) phylogeny of native, non-hybrid ferns in
Japan in newick format. Tree dated using treePL v1.0 [@Smith2012] with `r n_fossil_points` 
fossil calibration points after @Testo2016a. Values at nodes
are bootstrap support values calculated using 1000 replicates of ultrafast
bootstrap [@Nguyen2015]. Clades consisting of identical OTUs not collapsed.

Number of tips: `r ape::Ntip(japan_fern_tree_uncollapsed_dryad)`

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_fern_tree_uncollapsed_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

```{r japan-fern-dryad-file}
japan_fern_tree_dryad <- ape::read.tree(here::here(japan_fern_tree_dryad_file))
```

`r fs::path_file(japan_fern_tree_dryad_file)` (contained in "results.zip"):
Maximim-likelihood, ultrametric (dated) phylogeny of native, non-hybrid ferns in
Japan in newick format. Tree dated using treePL v1.0 [@Smith2012] with `r n_fossil_points` 
fossil calibration points after @Testo2016a. Values at nodes
are bootstrap support values calculated using 1000 replicates of ultrafast
bootstrap [@Nguyen2015]. Clades consisting of identical sequences have been
collapsed to polytomies with zero branch length between OTUs.

Number of tips: `r ape::Ntip(japan_fern_tree_dryad)`

Missing data codes: No missing data.

MD5 checksum: `r tools::md5sum(here::here(japan_fern_tree_dryad_file))`

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

CHANGE LOG

\-\-\-

2021-09-09

Generate this README file.

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

References
