---
title: "Exploring dimensions of biodiversity in Japanese ferns"
bibliography: "main_library.bib"
mainfont: Roboto-Regular.ttf
fontsize: 11pt
mainfontoptions: 
- BoldFont=Roboto-Bold.ttf
- ItalicFont=Roboto-Italic.ttf
- BoldItalicFont=Roboto-BoldItalic.ttf
output: 
  bookdown::pdf_document2:
    citation_package: none # Use pandoc default, otherwise csl can't be used for reference formatting
    toc: no
    number_sections: no
    fig_caption: yes
    keep_tex: no
    latex_engine: xelatex
    pandoc_args: [ "--csl", "journal-of-biogeography.csl"]
    includes:
      in_header: newphyt.sty
header-includes: 
  \usepackage{float}
  \makeatletter\renewcommand*{\fps@figure}{H}\makeatother
editor_options: 
  chunk_output_type: console
# nocite: | 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, results = "hide", cache = FALSE)
```

```{r load-data, cache = FALSE}
# NOTE: This Rmd is meant to be knit as part of the drake plan (R/plan.R).
# If using interactively, run `source("_drake.R")` first.

loadd(list = c(
  "occ_point_data_ferns",
  "occ_point_data_ferns_unfiltered",
  "ses_phy_ferns",
  "ses_phy_ferns_endemic",
  "ses_traits_ferns",
  "shape_ferns",
  "k_taxonomy",
  "k_phylogeny",
  "regions_taxonomy",
  "regions_phylogeny"
),
cache = ja_fern_cache)

phylo_cache <- new_cache(here::here("phylo_cache"))

loadd(list = c(
  "plastome_calibration_dates",
  "plastome_tree"
),
cache = phylo_cache)

# Load captions
source(here::here("ms/captions.R"))
```

<!-- Remove page numbers -->
\pagenumbering{gobble}

<!-- Set left justification -->
\raggedright

<!-- Don't indent -->
\setlength{\parindent}{0ex}

Joel H. Nitta^1^, Brent D. Mishler^2^, Wataru Iwasaki^1^, and Atsushi Ebihara^3^

^1^Department of Biological Sciences, Graduate School of Science, The University of Tokyo, 2-11-16 Yayoi, Bunkyo-ku, Tokyo 113-0032, Japan

^2^University and Jepson Herbaria, University of California, Berkeley, CA 94720‐2465, USA

^3^Department of Botany, National Museum of Nature and Science, 4-1-1 Amakubo, Tsukuba-shi, Ibaraki, 305-0005, Japan  

Author for correspondence:  
*Joel H. Nitta*  
*Email: joelnitta@gmail.com*

\clearpage

## Abstract

The ferns of Japan are one of the best-characterized regional floras in the world due to decades of intensive collection and taxonomic revision. We combined phylogenetic, functional, and environmental datasets to explore the spatial distribution of biodiversity in the ferns of Japan.

**Key words:** biodiversity, biogeography, phylogenetic diversity. 

\clearpage

## Introduction

Characterizing the spatial distribution of biodiversity is a major goal of evolutionary biology with two equally important aspects: to understand the processes generating it, and to conserve it. The vast majority of studies seeking to understand the spatial distribution of biodiversity focus on species richness. Hotspots, or areas of exceptionally high richness, have received particular attention, resulting in a widely-recognized set of 34 terrestrial hotspots and motivating conservation strategies to preserve them. However, richness alone cannot provide a complete picture of biodiversity. All organisms are related by descent from a common ancestor, and these evolutionary relationships must be taken into account to obtain a full understanding of biodiversity.

Presumably, the main reason phylogeny has not been taken into account more prominently in biodiversity studies is because data (DNA sequences) and analytical tools have only become available relatively recently. These tools and datasets now make it possible to analyze other dimensions of biodiversity, such as phylogenetic diversity and phylogenetic endemism. Better ways to characterize biological regions ("bioregions") are also available, rather than relying on ad-hoc characterizations. Furthermore, incorporating these two frameworks---the categorization of areas into phyloregions with analysis of over/under dispersion of biodiversity---can provide particularly powerful insights into the processes structuring biodiversity and suggest conservation priorities.

However, a comprehensive understanding of the relationships between richness and other metrics requires densely sampled data or can be lead astray.

Ferns of Japan are excellent model system because they have been the target of intense botanical interest for several decades and are very densely sampled. Japan is situated along a latitudinal gradient spanning highly season temperate areas in the north to subtropical, mostly aseasonal islands in the south.

Here we leverage this exceptionally rich dataset to analyze the geographic distribution of biodiversity in detail. We ask the following questions in our study system, the ferns of Japan: 1. How is biodiversity distributed? 2. How is biodiversity structured? 3. How well is biodiversity protected?

## Materials and Methods

### Study Site

Our data include all ferns known to occur within the political boundaries of Japan (20° to 45° N, 122° to 153° E). Japan consists of four main islands and thousands of smaller ones. The climate is subtropical in the south and highly seasonal in other parts of the country.

### Occurrence data

We used a list of all non-hybrid fern specimens housed at herbaria in Japan. We chose to use this over other sources (e.g., GBIF), because of its high taxonomic quality: all identifications are standardized to a common taxonomy, the Japan Green List v. 1.01 (http://www.rdplants.org/gl/). Furthermore, the overall sampling completeness of this dataset is extremely high (`r s_figure("sampling-curve")`). The original list comprises `r nrow(occ_point_data_ferns_unfiltered) %>% scales::number(big.mark = ",")` specimens representing `r occ_point_data_ferns_unfiltered %>% pull(taxon) %>% n_distinct()` taxa (species, subspecies, and varieties). We filtered out any occurrences not within a map of standard grid squares defined by the Statistics Bureau of Japan (http://www.stat.go.jp/english/data/mesh/05.html). The final cleaned occurrence dataset included `r nrow(occ_point_data_ferns) %>% scales::number(big.mark = ",")` specimens (`r occ_point_data_ferns %>% pull(taxon) %>% n_distinct()` taxa).

A necessary step in any analysis of biodiversity is to set the grain size used to define communities. This decision must consider that smaller grain size is needed to detect environmental effects, while larger grain size is needed to ensure adequate species sampling. We tested using 0.1, 0.2, 0.3, and 0.4 degree grid-cells for defining communities (co-occurring species). At each size, species occurrences were converted into a presence-absence community matrix (a species was considered present if at least one specimen was recorded in that grid cell). We calculated sampling redundance (1 - richness/number of specimens) [@Garcillan2003] to quantify adequacy of sampling. Preliminary analysis indicated that 0.2 degree grid-cells are optimal for our dataset (`r s_figure("grain-size")`). We filtered out any cells with redundancy < 0.1 (mostly cells along the coast with very low land area) and used this for subsequent analyses.

### Trait data

We used traits originally compiled for identification of ferns and lycophytes of Japan [@Ebihara2016b; @Ebihara2017; @Ebihara2019b]. Quantitative traits were measured on 10 randomly selected specimens per species, and qualitative traits determined by observing voucher specimens. All qualitative traits in binary format; e.g., a single trait with three states is formatted as three binary traits. Quantitative traits were measured in fertile (i.e., spore-bearing) and sterile fronds separately. Quantitative traits include aspects of leaf size (e.g., frond length, frond width, stipe length, etc.). Qualitative traits include aspects of leaf morphology (e.g., margin shape, leaf color, indusia shape, spore arrangement, etc.).

### Phylogenetic analysis

Sequences for the plastid, coding *rbcL* gene are available for ca. 97% of the Japanese fern flora [@Ebihara2019b]. However, using this dataset alone for phylogeny suffers two drawbacks: it is not densely sampled enough to resolve many internal nodes, and it lacks many lineages that have fossils availabsle for molecular dating. Because one goal of the study is to understand the distribution of paleo- vs. neo-endemic areas (defined in units of time), we require an ultrametric tree. To obtain a maximally sampled, robust phylogenetic tree, we combined the *rbcL* data of @Ebihara2019b with a globally sampled plastid dataset including all fern sequences on GenBank for four plastid genes (*rbcL*, *atpB*, *aptA*, and *rps4*) and 66 other coding, single-copy plastid genes extracted from 100 complete fern plastomes (Nitta et al., in prep) (`r ape::Ntip(plastome_tree) %>% scales::number(big.mark = ",")` OTUs total). We conducted phylogenetic analysis with IQTREE [@Nguyen2015] and molecular dating with treePL [@Smith2012] using `r nrow(plastome_calibration_dates)` fossil calibration points after @Testo2016a. We then trimmed the tree to Japanese taxa only and used this for all further analyses.  

### Analysis of biodiversity and endemism

We measured taxonomic diversity using taxonomic richness.

We measured phylogenetic diversity using the mean phylogenetic distance between all co-occurring taxa in a community (MPD) [@Webb2000; @Webb2002]. We calculated distances from the phylogenetic tree using the \function{cophenetic} function in R v. 4.0.0 [@RCoreTeam2019] and MPD using the \function{mpd} function in the R \package{picante}.

For morphological diversity, we first calculated morphological distances using Gower's method [@Gower1971] on the trait data with the \function{gowdis} function in the R package \package{FD} [@Laliberte2011]. We weighted quantitative and qualitative traits equally, weighted each binary component trait of each qualitative trait equally, and log-transformed and scaled quantitative traits prior to calculating distances. We then used the morphological distances as input into the \function{mpd} function to calculate mean morphological distance (MMD).

We measured phylogenetic endemism (PE) [@Rosauer2009], which is the sum of all branches at a site inversely weighted by their range, using the \function{phylo{\_}endemism} function in the R package \package{phyloregion}. 

Observed values of MPD, MMD, and PE are known to be closely related to species richness, and tend to saturate at high values of richness. Therefore, we determined whether the observed metrics are more or less diverse than random for a given number of species by calculating their standard effect sizes (SES), defined as the difference between the observed value and the mean of a pool of random values, divided by the standard deviation of the random values. We generated 999 random communities using the "independent swap" algorithm of @Gotelli2000, which conserves richness per site and species occurrence frequencies while randomizing species identities, then calculated SES of each metric.

We tested whether endemic areas have an over-representation of old (paleoendemic) or new (neoendemic) lineages using categorical analysis of neo- and paleo-endemism (CANAPE), which involves measuring PE with an alternate tree where all branch lengths are equal, then comparing the ratio between raw PE and alternative PE (RPE) [@Mishler2014]. Low RPE indicates short branchlengths (neoendemism) and high RPE indicates long branchlengths (paleoendemism). Since all of these measures are affected by species richness, categorization of endemism is done using the rank of each observed variable to those calculated for the null distribution. We used the same set of null communities as for SES of PE, and carried out CANAPE using custom R scripts.

It is possible that endemism may be affected by species that are actually wide-ranging, but only have a small range in Japan (the "border effect"). This is especially expected for southern, sub-tropical islands, which harbor many tropical species with small ranges in Japan but wider ranges outside of the country. To account for this, we also calculated all endemism-related metrics using a dataset including only species endemic to Japan.

### Analysis of bioregions

We analyzed bioregions using the framework of @Daru2020 and their R package \package{phyloregion}. Briefly, this involves first calculating beta diversity (change in composition between communities) using either species names (taxonomic bioregions) or phylogenetic distances (phylogenetic bioregions). For taxonomic bioregions, we calculated the turnover in species between sites due to species replacement [@Baselga2012] with the Sørensen index. For phylogenetic bioregions, we calculated phylogenetic dissimilarities based on the PhyloSor index [@Bryant2008]. The distances are used to construct a dendrogram, which is then split into into *k* bioregions. As the optimal number of bioregions cannot be known *a-priori*, we tested values of *k* from 1 to 20. \package{phyloregion} provides a function to automatically select *k* from a range of test values, but we found the results to be unreliable. We therefore evaluated the explained variance at each value of *k*, then selected *k* that provided the greatest increase in variance relative to the previous value (`r s_figure("k-plot")`).

### Data accessibility

Code to replicate all analyses, figures, and this manuscript are available at https://github.com/joelnitta/japan_ferns_biogeo. Raw and processed data are openly available from the Dryad repository at http://doi.org/ADD_DOI_HERE_WHEN_ASSIGNED.

## Results

### Taxonomic, phylogenetic, and morphological diversity

Species richness is highest in southern Kyushu and on Yakushima Island, and lowest on Hokkaido and low-lying coastal areas (`r figure("diversity")`A). Sampling redundancy is generally high throughout the country, with a peak just north of Tokyo, but lower in Hokkaido and some coastal areas and small islands (`r figure("diversity")`B). Raw MPD is mostly uniform throughout (`r figure("diversity")`C). SES of MPD shows clustering in some parts of central Honshu, with overdispersion both in the northern half of the country and from southern Kyushu to the south (`r figure("diversity")`D). Morphological diversity shows a similar pattern to phylogenetic diversity, but with more extreme overdispersion and clustering (`r figure("diversity")`F).

### Phylogenetic endemism

Phylogenetic endemism (PE) is extremeley low throughout except for in some of the southern islands (`r figure("endemism")`A). SES of PE is high starting in southern Kyushu and southwards, including the Ogasawara islands (`r figure("endemism")`B). Results of CANAPE (Categorical Analysis of Palaeo and Neo Endemism) indicate hotspots of mixed endemism in the southern islands, but no neo- and very little paleo-endemism (`r figure("endemism")`C). When we analyzed patterns of endemicity with a reduced dataset including only taxa endemic to Japan to account for the border effect, patterns were generally similar to the full dataset, indicating that the southern islands are genuine hotspots of endemicity (`r s_figure("endemism-restricted")`).

### Bioregions

Analysis of biogeographic regions using taxonomy (species names) resulted in the vast majority of grid cells classified into four regions: Cluster 1 (Hokkaido and northern Honshu), Cluster 2 (southern Honshu), Cluster 3 (Okinawa), and Cluster 4 (Ogasawara Islands) (`r figure("bioregions")`A). Using phylogenetic distances produced similar results, but Okinawa and the Ogasawara Islands merged into a single cluster (`r figure("bioregions")`B).

### Relationships between metrics

Raw values of MPD and MMD showed high variance and low species richness, then merged towards a single value at high species richness, as expected (`r s_figure("raw-biplot")`K, O). Raw PE was linearly correlated with species richness (`r s_figure("raw-biplot")`M). Both raw and SES values of MPD and MMD were highly correlated (`r s_figure("raw-biplot")`D, `r s_figure("ses-biplot")`D). SES values of MPD, MMD, and PE were all positively correlated (`r s_figure("ses-biplot")`). None of the SES values were correlated with richness (`r s_figure("ses-biplot")`).

## Discussion

TBD

## Acknowledgements

## Author Contribution

## References
<!-- Insert <div id="refs"></div> to place the refs here instead of at the end -->
<div id="refs"></div>

\clearpage

<!-- Figures -->

```{r plot-prep}
tax_regions_data <- regions_taxonomy %>%
  mutate(taxonomic_cluster = as.factor(cluster) %>% fct_infreq %>% as.numeric %>% as.factor) %>%
  select(grids, taxonomic_cluster)

phy_regions_data <- regions_phylogeny %>%
  mutate(phylo_cluster = as.factor(cluster) %>% fct_infreq %>% as.numeric %>% as.factor) %>%
  select(grids, phylo_cluster)

ses_div_ferns_spatial <-
  shape_ferns %>%
  left_join(ses_phy_ferns, by = c(grids = "site")) %>%
  left_join(ses_traits_ferns, by = c(grids = "site")) %>%
  left_join(tax_regions_data, by = "grids") %>%
  left_join(phy_regions_data, by = "grids")

ses_div_ferns_endemic_spatial <-
shape_ferns %>%
  left_join(ses_phy_ferns_endemic, by = c(grids = "site"))

map_theme <- theme(
    panel.grid.major = element_line(color = "grey95", size = 0.1),
    panel.background = element_rect(fill = "grey60"),
    axis.ticks = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.7)
  )

map_theme_light <- theme(
    panel.grid.major = element_line(color = "white", size = 0.1),
    panel.background = element_rect(fill = "grey90"),
    axis.ticks = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.7)
  )

```

```{r richness, width = 7, height = 4}
# Richness
a <- ggplot(ses_div_ferns_spatial, aes(fill = richness)) +
  geom_sf(color = "transparent") +
  scale_fill_scico(palette = "lajolla") + 
  labs(fill = "Richness") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme

# Sampling rendundancy
b <- ggplot(ses_div_ferns_spatial, aes(fill = redundancy)) +
  geom_sf(color = "transparent") +
  scale_fill_scico(palette = "lajolla") + 
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  labs(fill = "Redundancy") +
  map_theme +
  theme(
    axis.text.y = element_blank()
  )

a + b + 
  plot_layout(ncol = 2) +
  plot_annotation(tag_levels = 'A')

ggsave(result_file(figure("richness"), "pdf"), width = 7, height = 4, units = "in")
```

**`r figure("richness")`**. `r figure_cap("diversity")` **A**, richness. **B**, sampling redundancy.

```{r diversity, fig.width = 7, fig.height = 9}
# PD
a <- ggplot(ses_div_ferns_spatial, aes(fill = pd_obs)) +
  geom_sf(color = "transparent") +
  labs(fill = "PD") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  scale_fill_scico(palette = "lajolla") + 
  map_theme +
  theme(
    axis.text.x = element_blank()
  )

# SES of PD
b <- ggplot(ses_div_ferns_spatial, aes(fill = pd_obs_z)) +
  geom_sf(color = "transparent") +
  labs(fill = "SES PD") +
  scale_fill_scico(
    palette = "romaO",
    na.value="grey80",
    direction = -1,
    limits = c(
      -get_limit(ses_div_ferns_spatial, pd_obs_z, "abs"),
      get_limit(ses_div_ferns_spatial, pd_obs_z, "abs")
    )) +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank()
  )

# RPD
c <- ggplot(ses_div_ferns_spatial, aes(fill = rpd_obs)) +
  geom_sf(color = "transparent") +
  labs(fill = "RPD") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  scale_fill_scico(palette = "lajolla") + 
  map_theme +
  theme(
    axis.text.x = element_blank()
  )

# SES of RPD
d <- ggplot(ses_div_ferns_spatial, aes(fill = rpd_obs_z)) +
  geom_sf(color = "transparent") +
  labs(fill = "SES RPD") +
  scale_fill_scico(
    palette = "roma",
    na.value="grey80",
    direction = -1,
    limits = c(
      -get_limit(ses_div_ferns_spatial, rpd_obs_z, "abs"),
      get_limit(ses_div_ferns_spatial, rpd_obs_z, "abs")
    )) +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  # highlight only significant cells
  # gghighlight(pd_obs_p > 0.975 | pd_obs_p < 0.025) +
  map_theme +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank()
  )

# FD
e <- ggplot(ses_div_ferns_spatial, aes(fill = fd_obs)) +
  geom_sf(color = "transparent") +
  labs(fill = "FD") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  scale_fill_scico(palette = "lajolla") + 
  map_theme +
  theme(
    axis.text.x = element_blank()
  )

# SES of FD
f <- ggplot(ses_div_ferns_spatial, aes(fill = fd_obs_z)) +
  geom_sf(color = "transparent") +
  labs(fill = "SES FD") +
  scale_fill_scico(
    palette = "roma",
    na.value="grey80",
    direction = -1,
    limits = c(
      -get_limit(ses_div_ferns_spatial, fd_obs_z, "abs"),
      get_limit(ses_div_ferns_spatial, fd_obs_z, "abs")
    )) +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  # highlight only significant cells
  # gghighlight(mpd_morph_obs_p > 0.975 | mpd_morph_obs_p < 0.025) +
  map_theme +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank()
  )

# RFD
g <- ggplot(ses_div_ferns_spatial, aes(fill = rfd_obs)) +
  geom_sf(color = "transparent") +
  labs(fill = "RFD") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  scale_fill_scico(palette = "lajolla") + 
  map_theme

# SES of RFD
h <- ggplot(ses_div_ferns_spatial, aes(fill = rfd_obs_z)) +
  geom_sf(color = "transparent") +
  labs(fill = "SES RFD") +
  scale_fill_scico(
    palette = "roma",
    na.value="grey80",
    direction = -1,
    limits = c(
      -get_limit(ses_div_ferns_spatial, rfd_obs_z, "abs"),
      get_limit(ses_div_ferns_spatial, rfd_obs_z, "abs")
    )) +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  # highlight only significant cells
  # gghighlight(rfd_obs_p > 0.975 | rfd_obs_p < 0.025) +
  map_theme +
  theme(
    axis.text.y = element_blank()
  )

a + b + c + d + e + f + g + h +
  plot_layout(ncol = 2) +
  plot_annotation(tag_levels = 'A')

ggsave(result_file(figure("diversity"), "pdf"), width = 7, height = 9, units = "in")
```

**`r figure("diversity")`**. `r figure_cap("diversity")` **A**, Phylogenetic diversity (PD). **B**, Standard effect size (SES) of PD. **C**, Relative phylogenetic diversity (RPD). **D**, SES of RPD. **E**, Functional diversity (FD). **F**, SES of FD **G**, Relative functional diversity (RFD) **H**, SES of RFD. 

\clearpage

```{r endemism, fig.width = 7, fig.height = 8}
# PE
a <- ggplot(ses_div_ferns_spatial, aes(fill = pe_obs)) +
  geom_sf(color = "transparent") +
  scale_fill_scico(palette = "lajolla") + 
  labs(fill = "PE") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme +
  theme(
    axis.text.x = element_blank()
  )

# PE rank
b <- ggplot(ses_div_ferns_spatial, aes(fill = pe_obs_rank)) +
  geom_sf(color = "transparent") +
  labs(fill = "Rank PE") +
  scale_fill_scico(palette = "lajolla") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank()
  )

# RPE
c <- ggplot(ses_div_ferns_spatial, aes(fill = rpe_obs)) +
  geom_sf(color = "transparent") +
  scale_fill_scico(palette = "lajolla") +
  labs(fill = "RPE") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme +
  theme(
    axis.text.x = element_blank()
  )

# SES RPE
d <- ggplot(ses_div_ferns_spatial, aes(fill = rpe_obs_z)) +
  geom_sf(color = "transparent") +
  labs(fill = "SES RPE") +
  scale_fill_scico(
    palette = "roma",
    na.value="grey80",
    direction = -1,
    limits = c(
      -get_limit(ses_div_ferns_spatial, rpe_obs_z, "abs"),
      get_limit(ses_div_ferns_spatial, rpe_obs_z, "abs")
    )) +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme +
  theme(
    axis.text.y = element_blank()
  )

# CANAPE, full dataset
# set color scheme
canape_cols <- c(
  "neo" = "#E41A1C", # Color brewer set1 red
  "paleo" = "#377EB8", # Color brewer blue
  "mixed" = "#984EA3", # Color brewer purple
  "none" = "#FFFFCC" # Scico lajolla faintest color
)

e <- 
  ses_div_ferns_spatial %>%
  mutate(endem_type = replace_na(endem_type, "none")) %>%
  ggplot(aes(fill = endem_type)) +
  geom_sf(color = "transparent") +
  labs(fill = "Endemism\ntype") +
  scale_fill_manual(values = canape_cols, breaks = c("neo", "paleo", "mixed", "none"), labels = c("Neo", "Paleo", "Mixed", "Not endemic")) + 
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme_light

a + b + c + d + e +
  plot_layout(ncol = 2) +
  plot_annotation(tag_levels = 'A')

ggsave(result_file(figure("endemism"), "pdf"), width = 7, height = 8, units = "in")
```

**`r figure("endemism")`**. `r figure_cap("endemism")` **A**, phylogenetic endemism (PE). **B**, Standard effect size (SES) of PE. **C**, relative phylogenetic endemism (RPE). **D**, SES of RPE. **E**, Categorical analysis of neo- and paleo-endemism (CANAPE).

\clearpage

```{r bioregions, fig.width = 7, fig.height = 3}
# Taxonomic bioregions
a <- ggplot(ses_div_ferns_spatial, aes(fill = taxonomic_cluster)) +
  geom_sf(color = "transparent") +
  scale_fill_carto_d(name = "Cluster", type = "qualitative", palette = "Bold") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme_light

# Phylogenetic bioregions
b <- ggplot(ses_div_ferns_spatial, aes(fill = phylo_cluster)) +
  geom_sf(color = "transparent") +
  labs(fill = "Cluster") +
  scale_fill_carto_d(name = "Cluster", type = "qualitative", palette = "Bold") +
  scale_y_continuous(labels = scales::label_number(suffix = "\u00b0N", accuracy = 1)) +
  scale_x_continuous(labels = scales::label_number(suffix = "\u00b0E", accuracy = 1)) +
  map_theme_light

a + b +
  plot_layout(ncol = 2) +
  plot_annotation(tag_levels = 'A')

ggsave(result_file(figure("bioregions"), "pdf"), width = 7, height = 3, units = "in")
```

**`r figure("bioregions")`**. `r figure_cap("bioregions")` **A**, taxonomic bioregions. **B**, phylogenetic bioregions. Bioregions determined by clustering taxonomic (Sørensen) or phylogenetic (PhyloSor) distances into *k* = 8 or 7 groups, respectively, using the R package \package{phyloregion}.

\clearpage

```{r diversity-by-cluster, fig.width = 7, fig.height = 7}
data <-
ses_div_ferns_spatial %>%
  filter(taxonomic_cluster %in% c(1,2,3,4)) %>%
  as_tibble %>%
  select(pd_obs_z, rpd_obs_z, fd_obs_z, rfd_obs_z, pe_obs_rank, taxonomic_cluster)

cluster_cols <- carto_pal(8, "Bold")[1:4] %>%
  set_names(1:4)

a <- ggplot(data, aes(x = taxonomic_cluster, y = pd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES PD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

b <- ggplot(data, aes(x = taxonomic_cluster, y = rpd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RPD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

c <- ggplot(data, aes(x = taxonomic_cluster, y = fd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES FD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

d <- ggplot(data, aes(x = taxonomic_cluster, y = rfd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RFD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.title.x = element_blank()
  )

e <- ggplot(data, aes(x = taxonomic_cluster, y = pe_obs_rank)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "Rank PE",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    
  )

a + b + c + d + e +
  plot_layout(ncol = 2) +
plot_annotation(
  tag_levels = 'A'
) &
  theme(legend.position = "none")
  
ggsave(result_file(figure("div-by-tax-region"), "pdf"), width = 7, height = 7, units = "in")

```

**`r figure("div-by-tax-region")`**. `r figure_cap("div-by-tax-region")`. **A**, Standard effect size (SES) of phylogenetic diversity (PD). **B**, SES of relative PD (RPD). **C**, SES of functional diversity (FD). **D**, SES of relative FD (RFD). **E**, rank of observed phylogenetic endemism (PE) out of 999 random communities. Only top four clusters with the most sites shown.