---
mainfont: Roboto-Regular.ttf
fontsize: 11pt
mainfontoptions:
  - BoldFont=Roboto-Bold.ttf
  - ItalicFont=Roboto-Italic.ttf
  - BoldItalicFont=Roboto-BoldItalic.ttf
output:
  bookdown::word_document2:
    pandoc_args: [ "--csl", "american-journal-of-botany.csl"]
    toc: no
    number_sections: no
    reference_docx: american-journal-of-botany.docx
  bookdown::pdf_document2:
    citation_package: default
    toc: no
    number_sections: no
    fig_caption: no
    latex_engine: xelatex
    pandoc_args: [ "--csl", "american-journal-of-botany.csl"]
params:
  doc_type: doc
editor_options:
  chunk_output_type: console
---

---
bibliography:
  - `r here::here("ms/references.yaml")`
  - `r here::here("ms/references_other.bib")`
---

```{r manual, include = FALSE, evaluate = FALSE}
# Run these lines if rendering outside of targets plan
library(targets)
library(tarchetypes)
source(here::here("R/packages.R"))
source(here::here("R/functions.R"))
```

```{r setup, include = FALSE, cache = FALSE}
# Load libraries used for rendering Rmds
library(flextable)
library(kableExtra)
library(ftExtra)
library(scales)
library(patchwork)
library(RColorBrewer)
library(ggspatial) # for scalebars on maps

# Set knitr options
knitr::opts_chunk$set(
  echo = FALSE, message = FALSE, warning = FALSE,
  results = "hide", cache = FALSE)

# Load functions and variables only used when rendering Rmds
source(here::here("R/ms_functions.R"), local = TRUE)
```

```{r load-data}
# Load objects from targets workflow
tar_load(c(
  "aic_env_repro",
  "binary_sig_results",
  "biodiv_ferns_cent",
  "biodiv_ferns_spatial",
  "broad_alignment_list",
  "comm_ferns",
  "comm_ferns_full",
  "csl_file",
  "deer_range",
  "fern_traits",
  "green_list",
  "japan_fern_tree",
  "japan_map_points",
  "japan_pterido_tree",
  "japan_rbcL",
  "japan_shp",
  "k_phylogeny",
  "k_taxonomy",
  "lat_span_summary",
  "lrt_results",
  "model_fits",
  "model_params",
  "ms_functions",
  "occ_point_data_summary",
  "phy_sig_results",
  "plastome_calibration_dates",
  "plastome_tree",
  "ppgi",
  "protected_areas",
  "refs_other_yaml",
  "refs_yaml",
  "refs_other_bib",
  "repro_data",
  "signif_cells_protected_area",
  "signif_cells_deer_danger",
  "template_file",
  "traits_summary",
  "biodiv_ferns_spatial_lumped",
  "bioregion_cutoff"
))
```

<!-- Set left justification -->
\raggedright

<!-- Don't indent -->
\setlength{\parindent}{0ex}

RESEARCH ARTICLE

Short Title: Spatial phylogenetics of Japanese ferns

__Spatial phylogenetics of Japanese ferns: patterns, processes, and implications for conservation__

Joel H. Nitta^1^, Brent D. Mishler^2^, Wataru Iwasaki^1^, and Atsushi Ebihara^3^

```{r author-page-doc, results = "asis", echo = FALSE, eval = params$doc_type == "doc"}
# Author page, version for submission to journal
cat('
Manuscript received 22 September 2021; revision accepted 11 March 2022
')
```

^1^Department of Integrated Biosciences, Graduate School of Frontier Sciences, The University of Tokyo, Chiba, Japan

^2^University and Jepson Herbaria, and Department of Integrative Biology, University of California, Berkeley, CA, USA

^4^Department of Botany, National Museum of Nature and Science, Tsukuba, Japan

__Correspondence__

Joel H. Nitta, #601 Iwasaki Lab, Bioscience Bldg, The University of Tokyo, 5-1-5 Kashiwanoha, Kashiwa, Chiba 277-0882, Japan

Email: joelnitta@gmail.com

ORCID ID: 0000-0003-4719-7472

```{r author-page-doc-2, results = "asis", echo = FALSE, eval = params$doc_type == "doc"}
# Author page, version for submission to journal
cat('
__Citation__: Nitta, J. H., B. D. Mishler, W. Iwasaki, and A. Ebihara. 2022. Spatial phylogenetics of Japanese ferns: Patterns, processes, and implications for conservation. *American Journal of Botany* 109(5): XXXX

__DOI: XXXX__
')
```

\newpage

**Abstract**

```{r abstract-stats}
# Convert occ_point_data_summary dataframe to nested list for easier indexing
occ_sum <- occ_point_data_summary %>%
  # change dataset name to just "unfiltered" vs. "filtered"
  mutate(dataset = case_when(
    str_detect(dataset, "unfiltered") ~ "unfiltered",
    TRUE ~ "filtered")) %>%
  split(.$dataset) %>%
  map(., ~split(., .$variable))

# Count number of major bioregions after lumping minor regions into "Other"
maj_cluster_count <-
biodiv_ferns_spatial_lumped %>%
  select(grids, contains("cluster")) %>%
  st_drop_geometry() %>%
  as_tibble() %>%
  pivot_longer(names_to = "type", values_to = "cluster", -grids) %>%
  group_by(type) %>%
  count(cluster) %>%
  filter(cluster != "Other") %>%
  summarize(n = n_distinct(cluster), .groups = "drop") %>%
  # Convert to list for easy indexing
  split(.$type)
```

**Premise:** Biodiversity is often only measured with species richness; however, this measure ignores evolutionary history and is not sufficient for making conservation decisions.
Here, we characterize multiple facets and drivers of biodiversity to understand how these relate to bioregions and conservation status in the ferns of Japan.

**Methods:** We compiled a community data set of `r nrow(comm_ferns) %>% scales::number(big.mark = ",")` grid cells (20 km × 20 km each) including `r ncol(comm_ferns)` taxa based on > `r occ_sum$filtered$n_specimens$value %>% plyr::round_any(1000, f = floor) %>% number(big.mark = ",")` specimen records.
We combined the community data with a phylogeny and functional traits to analyze taxonomic, phylogenetic, and functional diversity and modeled biodiversity metrics in response to environmental factors and reproductive mode.
Hierarchical clustering was used to delimit bioregions.
Conservation status and threats were assessed by comparing the overlap of significantly diverse grid cells with conservation zones and range maps of native Japanese deer.

**Results:** Taxonomic richness was highest at mid latitudes.
Phylogenetic and functional diversity and phylogenetic endemism were highest in small southern islands.
Relative phylogenetic and functional diversity were high at high and low latitudes, and low at mid latitudes.
Grid cells were grouped into `r maj_cluster_count$phylo_cluster$n %>% english2` (phylogenetic) or `r maj_cluster_count$taxonomic_cluster$n %>% english2` (taxonomic) major bioregions.
Temperature and apomixis were identified as drivers of biodiversity patterns.
Conservation status was generally high for grid cells with significantly high biodiversity, but the threat due to herbivory by deer was greater for taxonomic richness than other metrics.

**Conclusions:** Our integrative approach reveals previously undetected patterns and drivers of biodiversity in the ferns of Japan.
Future conservation efforts should recognize that threats can vary by biodiversity metric and consider multiple metrics when establishing conservation priorities.

<!-- may include title words -->
**Keywords:** apomixis, biodiversity, biogeography, CANAPE, conservation, ferns, Japan, phylogenetic diversity.

\newpage

```{r intro-stats}
# Make list of all native ferns in Japan (based off Green List)
# *including* hybrids
fern_list <-
  green_list %>%
  mutate(
    genus = str_split(taxon, "_") %>%
      map_chr(1)
  ) %>%
  left_join(ppgi, by = "genus") %>%
  assert(not_na, class) %>%
  filter(class == "Polypodiopsida")

fern_list_non_hybrid <-
  fern_list %>%
  filter(hybrid == FALSE) %>%
  mutate(
    gn_parse_tidy(.$scientific_name) %>%
      select(taxon = canonicalsimple)
  ) %>%
  mutate(taxon = str_replace_all(taxon, " ", "_"))

# Caculate number of native, non-hybrid fern taxa
total_fern_taxa <- fern_list_non_hybrid %>%
  pull(taxon) %>%
  n_distinct()

# Tally number of non-hybrid ferns with known repro. mode
repro_count <-
  repro_data %>%
  # Filter to only non-hybrid ferns
  inner_join(
    select(fern_list_non_hybrid, taxon_id),
    by = "taxon_id"
  ) %>%
  count(reproductive_mode) %>%
  mutate(mode = if_else(reproductive_mode == "unknown", "unknown", "known")) %>%
  group_by(mode) %>%
  summarize(total = sum(n)) %>%
  split(.$mode)

# Calculate percent of species with known reproductive mode
repro_count$known$percent <-
  repro_count$known$total %>%
  magrittr::divide_by(repro_count$unknown$total + repro_count$known$total) %>%
  percent(accuracy = 0.1)
```

Characterizing the spatial distribution of biodiversity is a major goal of evolutionary biology with two equally important aspects: to understand the processes generating it and to conserve it.
Until recently, the vast majority of studies seeking to understand the spatial distribution of biodiversity have focused on species richness.
Hotspots, or areas of exceptionally high species richness or endemism, have received particular attention, resulting in a widely recognized set of 36 terrestrial hotspots [@Myers2000; @Noss2016] and motivating conservation strategies to preserve them [@Margules2000; @Brooks2006].
However, species richness alone cannot provide a complete picture of biodiversity [@Miller2018].
All organisms are related to a greater or lesser degree by descent from a common ancestor, and these evolutionary relationships must be taken into account to obtain a full understanding of the biodiversity present in an area.

Presumably, phylogeny has not been taken into account more prominently in biodiversity studies mainly because the necessary data (DNA sequences and georeferenced occurrence records) and analytic tools have only become available relatively recently [@Soltis2016a; @Folk2021].
These datasets and tools now make it possible to analyze other dimensions of biodiversity such as phylogenetic diversity [@Faith1992] and phylogenetic endemism [@Rosauer2009].
Better ways to characterize biological regions ("bioregions"; `r ie`, areas defined by their taxonomic composition or evolutionary history) are also available [@Laffan2016; @White2019; @Daru2020a], rather than relying on ad hoc characterizations.
Furthermore, incorporating these two frameworks---the categorization of areas into phyloregions with analysis of over/under dispersion of biodiversity---can provide powerful insights into the processes structuring biodiversity and suggest conservation priorities.

A comprehensive understanding of the relationships between richness and other metrics requires densely sampled data, but such datasets are rare on the regional (country) scale.
The ferns of Japan are excellent model system because they have been the target of intense botanical interest for several decades and are densely sampled [reviewed by @Ebihara2019b].
The ferns of Japan include `r total_fern_taxa` native, nonhybrid taxa (including species and varieties) and hundreds of hybrids [nothotaxa\; @Ebihara2019b].
The availability of detailed distribution data [distribution maps at the ca. 10 km scale for nearly all species\; @Ebihara2016b; @Ebihara2017], trait data [multiple quantitative and qualitative traits compiled for identification of nearly all species\; @Ebihara2019b], and DNA sequences for > 97% of species [@Ebihara2010; @Ebihara2019b\; all coverage statistics exclude nothotaxa] make the ferns of Japan an ideal system for investigating the relationships between, and drivers of, multiple dimensions of biodiversity.

One particularly valuable characteristic of the Japanese fern flora is the availability of data on reproductive mode, which is known for `r repro_count$known$total` native fern taxa excluding hybrids [`r repro_count$known$percent`\; @Ebihara2019b].
Reproductive mode is likely to affect population-level genetic diversity [@Bengtsson2003] and thereby higher-level biodiversity [@Krueger-Hadfield2019].
In ferns, clonally reproducing apomictic species are often polyploid hybrids that share identical plastid genotypes among taxa and within populations [@Grusz2009; @Chao2012a; @Hori2014; @Hori2019].
Therefore, we expect high prevalence of apomictic species within a community to decrease phylogenetic diversity.

Japan is situated along a latitudinal gradient spanning temperate, seasonal areas in the north to subtropical, mostly aseasonal islands in the south.
Furthermore, it is a mountainous country with great variation in elevation on the larger islands.
Plant distributions often reflect physiological adaptations to climate [`r eg`, precipitation and temperature\; @Woodward1987].
Therefore, we expect the spatial distribution of biodiversity to be determined by climatic variation in addition to reproductive mode.

Here, we leveraged this exceptionally rich data set to analyze the geographic distribution of biodiversity in detail.
We asked the following questions in our study system, the ferns of Japan: 1. How is biodiversity distributed? 2. How is biodiversity structured? 3. What environmental and biological factors influence biodiversity? 4. How well is biodiversity protected?

## MATERIALS AND METHODS

All computational analyses were carried out in R v`r getRversion()` [@RCoreTeam2021] unless otherwise stated. The R package `r pack("targets")` was used to control analysis workflow [@Landau2021].

## Study Site

Japan (20°N to 45°N, 122°E to 153°E) consists of four main islands and thousands of smaller ones (`r figure("japan-map")`).
Most of the main islands have been in contact with the continent at various points in the past during periods of lower sea level, but the Ogasawara archipelago consists of oceanic islands that have never been connected to the mainland.
The climate is subtropical in the southern islands and temperate in other parts of the country.
Map data for Japan were downloaded from the Geospatial Information Authority of Japan under the Creative Commons Attribution License v4.0 (https://www.gsi.go.jp/kankyochiri/gm_japan_e.html).

## Occurrence data

We used a list of native, nonhybrid fern specimens deposited at herbaria in Japan to quantify occurrences [@Ebihara2016b; @Ebihara2017; @Ebihara2019b].
We chose to use this over other sources such as the Global Biodiversity Information Facility (GBIF) because of its high quality, which obviates many cleaning steps otherwise needed when working with large, publicly available datasets.
Furthermore, the overall sampling completeness of this dataset is also high (`r s_figure("sampling-curve")` in Appendix S1; see Supplemental Data with this article).
The original list comprises `r occ_sum$unfiltered$n_specimens$value %>% scales::number(big.mark = ",")` specimens representing `r occ_sum$unfiltered$n_taxa$value` terminal taxa at the rank of species, subspecies, form, or variety, excluding non-native taxa and nothotaxa.
We treated infraspecific taxa as distinct species during analysis and hereafter use "taxon" and "species" interchangeably to refer to entities at this rank, unless otherwise indicated.
All names are standardized to a common taxonomy, the Japan Green List v. 1.01 [@Ebihara2019c].

Occurrence records were georeferenced and thoroughly vetted as follows [@Ebihara2016b; @Ebihara2017; @Ebihara2019b].
For specimens that lacked latitude and longitude data, georeferencing was done by mapping collection site names to a set of standard ca. 10 km × 10 km grid squares defined by the Statistics Bureau of Japan (the "second-degree mesh"; http://www.stat.go.jp/english/data/mesh/05.html), and the centroid of the second-degree mesh cell used as the specimen location.
In the case that the collection site could not be mapped to a single second-degree mesh cell, it was excluded.
Next, occurrence maps were generated for each taxon showing presence or absence in each second-degree mesh cell.
In the case that a given taxon appeared insufficiently sampled (`r ie`, not present on the map where it would typically be expected to occur), AE and members of the Nippon Fernist Club (local botanists familiar with the ferns of Japan) searched for additional specimens, which were then added to the list.
The occurrence maps were iteratively refined until the vast majority of known Japanese ferns had been observed across their expected ranges; the resulting set of occurrence records can be considered accurate to ca. 10 km (the grain size of the second-degree mesh map).

We further cleaned the list before analysis by filtering out any occurrences not within the second-degree mesh and removing duplicate collections (`r occ_sum$filtered$n_specimens$value %>% scales::number(big.mark = ",")` specimens, `r occ_sum$filtered$n_taxa$value` taxa after filtering).
Given the high quality of our occurrence data and the fact that automated occurrence cleaning algorithms [`r eg`, CoordinateCleaner\; @Zizka2019] have the potential to erroneously exclude true occurrence points [`r ie`, false positives\; @Zizka2020], we chose not to apply additional automated cleaning steps to our data as is often done with occurrence records obtained from GBIF [`r eg`, @Rice2019a; @Suissa2021].

A necessary step in any analysis of biodiversity is to set the grain size used to accurately define communities (co-occurring species).
Here, one must consider that smaller grain size is needed to detect environmental effects, while larger grain size is needed to ensure adequate species sampling.
We created sets of 10-km, 20-km, 30-km, and 40-km grid cells covering Japan using a Mollweide equal-area projection (size refers to the length of the side of each square grid cell).
At each grain size, species occurrences were converted to a presence–absence community matrix (a species was considered present if at least one specimen was recorded in that grid cell).
We calculated sampling redundancy [1 - Richness/Number of specimens\; @Garcillan2003] to quantify adequacy of sampling.
Preliminary analysis indicated that 20-km grid cells are optimal for our data set; redundancy values greatly increased from 10 to 20 km, but much less improvement as grain size is increased beyond 20 km (`r s_figure("grain-size")`).
Although the grid cells are defined with equal area, actual land area of each grid cell varies due to coastlines.
Strictly filtering out all grid cells with less than complete land area would result in a large loss of data, as Japan has many coastlines and small islands.
Therefore, we instead filtered the grid cells by sampling completeness, removing any grid cells with redundancy < 0.1; this dataset was used for all subsequent analyses.
The R package `r pack("sf")` was used for all GIS analyses [@Pebesma2018].

## Morphological trait data

We used traits originally compiled for identification of ferns and lycophytes of Japan [@Ebihara2016b; @Ebihara2017], which were formatted to be used with Lucid software (https://www.lucidcentral.org/).
Continuous traits were measured on 10 randomly selected specimens per species, then four values per species were obtained from these measurements following the Lucid format: outside (`r ie`, outside the typical range) minimum, typical minimum, typical maximum, and outside maximum.
For species with dimorphic fertile (`r ie`, spore-bearing) and sterile fronds, fertile and sterile fronds were measured separately.
We took the mean of the typical minimum and maximum values to use as the species mean value.
Qualitative traits were scored by observing voucher specimens.
All qualitative traits were scored in binary format; `r eg`, a single trait with three states was formatted as three binary traits.
From the original trait list, we selected only traits with putative ecological function (`r table("traits-used")`) and excluded any traits that had Pearson correlation coefficient > 0.6 or fewer than three observations of a given trait state.

## Reproductive mode data

We used data compiled by @Ebihara2019b, which classifies each fern taxon as sexual, apomictic, mixed (both sexual and apomictic modes known from the same taxon), or unknown.
We calculated the percentage of apomictic taxa in each grid cell as the sum of apomictic and mixed taxa divided by total number of taxa (`r s_figure("envir")`B).

## Environmental data

We downloaded climate data at the 2.5 minute scale from the WorldClim database using the `r func("getData")` function in the R package `r pack("raster")` [@R-raster], and extracted four variables related to our research questions as follows: mean annual temperature (bio1; hereafter "temperature"), temperature seasonality (bio4), annual precipitation (bio12; hereafter "precipitation"), and precipitation seasonality (bio15) (`r s_figure("envir")`).
We identified the intersection of the climate data at the 2.5-minute scale with the 20-km grid cells, then calculated the mean of each climatic variable for each grid cell.
We calculated area as rolling mean area (km^2^) across 100-km latitudinal bands.
This approach reflects the idea that the amount of area with a given climate regime enclosing each grid cell may influence species richness of that cell [@Fahrig2013], and that using raw island areas would not be appropriate in this context because there are just four very large islands with many much smaller ones located close by (`r figure("japan-map")`).

## Phylogenetic analysis

```{r phylo-analysis-stats}
# Calculate stats related to Fern Tree of Life (FTOL) dataset

# Define outgroup taxa
plastome_og <-
  tibble(
    taxon = c(
      "Physcomitrium_patens",
      "Marchantia_polymorpha",
      "Anthoceros_angustus",
      "Isoetes_virginica",
      "Lycopodium_clavatum",
      "Selaginella_moellendorffii",
      "Sciadopitys_verticillata",
      "Ginkgo_biloba",
      "Kadsura_longipedunculata",
      "Trithuria_inconspicua",
      "Amborella_trichopoda",
      "Acorus_calamus",
      "Magnolia_tripetala",
      "Lamprocapnos_spectabilis",
      "Plumbago_auriculata",
      "Nicotiana_tabacum",
      "Liquidambar_formosana",
      "Sorbus_torminalis"),
    taxon_status = "outgroup")

# Define Sanger-sequenced genes
sanger_genes <- c("atpA", "atpB", "rbcL", "rps4")

# Make gene x taxon table for plastome alignment
plastome_sampling <-
  tibble(
    gene = names(broad_alignment_list),
    taxon = map(broad_alignment_list, rownames)
  ) %>%
  unnest(cols = taxon) %>%
  # Check that all Sanger genes are included
  verify(all(sanger_genes %in% gene)) %>%
  # Check that all outgroup taxa are included
  verify(all(plastome_og$taxon %in% taxon)) %>%
  # Categorize genes into Sanger or other
  mutate(data_type = if_else(
    gene %in% sanger_genes,
    "sanger",
    "full_plastome"
  ))

# Number of full plastome genes in FTOL dataset
n_plastome_genes <- plastome_sampling %>%
  filter(data_type == "full_plastome") %>%
  pull(gene) %>%
  n_distinct() %>%
  number(big.mark = ",")

# Number of fern taxa with full plastome sequences in FTOL dataset
n_plastome_fern_taxa <- plastome_sampling %>%
  filter(data_type == "full_plastome") %>%
  # Exclude outgroups
  anti_join(plastome_og, by = "taxon") %>%
  pull(taxon) %>%
  n_distinct() %>%
  number(big.mark = ",")

# Number of fern taxa in FTOL dataset
n_ftol_fern_taxa <- plastome_sampling %>%
  # Exclude outgroups
  anti_join(plastome_og, by = "taxon") %>%
  pull(taxon) %>%
  n_distinct() %>%
  number(big.mark = ",")

# Number of taxa in Japan fern rbcL dataset
n_japan_fern_taxa <- japan_rbcL %>%
  rownames %>%
  n_distinct() %>%
  number(big.mark = ",")

# Verify that tips in the plastome tree are accounted for
# by FTOL sampling + japan fern rbcL
tibble(taxon = plastome_tree$tip.label) %>%
  anti_join(select(plastome_sampling, taxon) %>% unique, by = "taxon") %>% # nolint
  anti_join(tibble(taxon = rownames(japan_rbcL)), by = "taxon") %>%
  verify(nrow(.) == 0, success_fun = success_logical)

# Number of fossils used for dating
n_fossil_points <- plastome_calibration_dates %>%
  filter(age_type != "fixed") %>%
  nrow()
```

Sequences for the plastid, coding *rbcL* gene are available for ca. 97% of the Japanese fern flora [@Ebihara2019b]; these sequences originate from samples that were identified using the same taxonomic system as the occurrence data [@Ebihara2010; @Ebihara2019b], so we used these preferentially over other data available on GenBank that might include misidentifications or different taxon concepts.
However, using this dataset alone for phylogenetic analysis suffers two drawbacks: it is not densely sampled enough to resolve some internal nodes, and it lacks many lineages that have fossils available for molecular dating.
Furthermore, use of community phylogenies generated by only by sampling the species present in the local community has been shown to produce spurious results in simulation studies [@Park2018].
Because one goal of the study is to understand the distribution of paleo- vs. neoendemic areas (defined in units of time), we require an ultrametric tree.
Therefore, to obtain a robust, ultrametric tree, we combined the *rbcL* data of @Ebihara2019b (`r n_japan_fern_taxa` taxa) with a globally sampled plastid dataset including all fern sequences on GenBank for four widely sequenced plastid genes (*atpA*, *atpB*, *rbcL*, and *rps4*; `r n_ftol_fern_taxa` taxa) and `r n_plastome_genes` other coding, single-copy plastid genes extracted from `r n_plastome_fern_taxa` complete fern plastomes [FTOL v0.0.1\; @Nitta2022b] (`r ape::Ntip(plastome_tree) %>% scales::number(big.mark = ",")` operational taxonomic units [OTUs] total, including outgroup taxa).
This gene sampling is comparable to a recent global fern phylogeny that resolved relationships across ca. 4,000 taxa using six plastid markers [@Testo2016a], and the addition of plastome data can be expected to increase support along the backbone.

We conducted maximum-likelihood phylogenetic analysis with `r software("IQ-TREE")` v1.6.12 [@Nguyen2015] under the GTR+I+G model (all sites concatenated into a single data matrix), and assessed node support with 1,000 ultrafast bootstrap replicates [@Minh2013].
Molecular dating was conducted with the program `r software("treePL")` v1.0 [@Smith2012] using `r n_fossil_points` fossil calibration points after @Testo2016a, with the exception of treating fossil *Kuylisporites* as belonging to crown *Cyathea*, not *Cyathea* + *Alsophila* [@Loiseau2020].
The root of the tree (most recent common ancestor of all land plants) was fixed to 475 mya [@Testo2016a].
We trimmed the resulting ultrametric tree to only Japanese taxa.
`r software("treePL")` requires branch lengths to be >0, and it sets extremely small branches to an arbitrary minimum positive length.
This behavior may artificially add branch length between taxa with identical DNA sequences, which occur multiple times in the Japanese fern *rbcL* dataset.
We converted any clades in the dated tree consisting of identical sequences to polytomies with the node at 0 million years ago (i.e., the present; Ma).
This tree was used for all subsequent analysis unless indicated otherwise.

## Phylogenetic signal

We tested for phylogenetic signal in continuous traits using two alternative metrics, Blomberg's *K* [@Blomberg2003] and Pagel's $\lambda$ [@Pagel1999].
*K* describes the ratio between the observed variance in a trait vs. the amount of variance expected under brownian motion (BM); $\lambda$ is a scaling parameter between 0 and 1 that transforms the phylogeny such that trait values fit those most closely expected under BM.
Both $\lambda$ and *K* = 1 under BM; relatively lower values indicate less signal (overdispersion) and higher values indicate more signal (clustering).
We measured *K* and $\lambda$ using the `r func("phylosig")` function in the R package `r pack("phytools")` [@Revell2012], which also calculates significance based on a randomization test for *K* and a likelihood ratio test for $\lambda$.
We tested for phylogenetic signal in qualitative (coded as binary) traits using Fritz and Purvis' *D* [@Fritz2010] with the `r func("phylo.d")` function in the R package `r pack("caper")` [@Orme2018], which also assesses significance by comparing the observed value to a null distribution of random values and a simulated distribution of values expected under BM.
Values of *D* range from 0 under BM to 1 under random evolution, but can exceed this range in cases of extreme clumping or overdispersion, respectively.
These methods for phylogenetic signal cannot use a tree with zero-length polytomies, so we used the original dated tree from treePL.

## Analysis of biodiversity

We measured taxonomic diversity using taxonomic richness (`r ie`, number of taxa in each community).

We measured phylogenetic diversity (PD) using the Faith's PD [the total branch length connecting all terminal taxa in a community, including the root of the phylogeny\; @Faith1992].
To detect areas with unusually long or short branches, we calculated relative phylogenetic diversity (RPD), which is a ratio of observed PD to PD measured on a tree where all branch lengths have been transformed to equal length [@Mishler2014].

We measured functional diversity (FD) using the method of @Petchey2002.
We first calculated functional distances using Gower's method [@Gower1971] on the trait data with the `r func("gowdis")` function in the R package `r pack("FD")` [@Laliberte2010].
We weighted continuous and qualitative traits equally, weighted each binary component trait of each qualitative trait equally, and log-transformed and scaled continuous traits prior to calculating distances.
We also tested alternative weighting schemes, but these produced very similar results (data not shown).
We then used the distances to build a trait dendrogram using the `r func("hclust")` function in the R package `r pack("stats")` [@RCoreTeam2021], and used the dendogram in place of a phylogenetic tree to calculate PD and RPD as described above.
These functional analogs of PD and RPD are hereafter referred to as FD (functional diversity) and RFD (relative functional diversity).

We measured phylogenetic endemism (PE) as the sum of all branches at a site inversely weighted by their range size [@Rosauer2009].

Observed values of PD, RPD, FD, RFD, and PE are known to be related to species richness; as more taxa are drawn at random without replacement from a given set, it becomes increasingly unlikely to draw a taxon that is distantly related (or functionally dissimilar) to the others.
Therefore, we determined whether the observed metrics are more or less diverse than random for a given number of taxa by conducting a randomization test.
We generated 999 random communities using the curveball algorithm of @Strona2014, which conserves richness per site and species occurrence frequencies while randomizing species identities, then compared the observed value to the distribution of random values.
Statistical significance was assessed using a one-tailed test for PE or a two-tailed test for other metrics with $\alpha$ = 0.5; observed values in the extreme 5% of the null distribution were considered significant.
We also calculated the standard effect size (SES) of each biodiversity metric.
SES measures how extreme the observed value (Obs) is relative to the null distribution (Null):

$SES = \frac{Obs - Mean(null)}{SD(Null)}$

We tested whether centers of endemism have an over-representation of old (paleoendemic) or new (neoendemic) lineages using categorical analysis of neo- and paleoendemism (CANAPE), which involves measuring PE with an alternate tree where all branch lengths are equal, then comparing the ratio between raw PE and alternative PE (RPE) [@Mishler2014].
Low RPE indicates short range-restricted branches (neoendemism) and high RPE indicates long range-restricted branches (paleoendemism).
Since all of these measures are affected by species richness, categorization of endemism is done using the rank of each observed variable relative to those calculated for the null distribution. All measures of biodiversity and CANAPE were carried out using the R package `r pack("canaper")` [@Nitta2022].

It is possible that endemism may be affected by species that are actually wide-ranging, but only have a small range in Japan (a "border effect").
This is especially expected for southern, subtropical islands, which harbor many tropical species with small ranges in Japan but wider ranges outside of the country.
To partially account for the border effect, we also calculated all endemism-related metrics using a data set including species found only in Japan.

## Spatial models

Two grid cells lacking environmental data, one grid cell occupied by a single species lacking reproductive mode data, and an outlier with 100% apomictic taxa due to extremely low species richness were excluded from modeling analyses (Appendix S2; see Supplemental Data with this article).

We tested for spatial autocorrelation in biodiversity metrics (richness, SES of PD, SES of FD, SES of RPD, SES of RFD) and independent variables (environmental variables and percentage apomictic taxa [hereafter, % apomictic taxa]) using Moran's I with the `r func("moran.mc")` function in the R package `r pack("spdep")` [@Bivand2013].
In all cases, significant spatial autocorrelation was detected (Moran's I > 0, *P* value of permutation test < 0.05) (`r s_table("moran")`).
Therefore, we used spatially explicit methods for modeling.

We checked for correlation between independent variables while taking into account spatial autocorrelation using a modified *t*-test with the `r func("modified.ttest")` function in the R package `r pack("SpatialPack")` [@SpatialPack2020].
This test indicated that temperature seasonality and % apomictic taxa were correlated with mean annual temperature (`r s_table("corr")`), so for the main modeling analysis we only included the following environmental variables: mean annual temperature, precipitation, and precipitation seasonality.

We constructed a linear mixed model for each biodiversity metric dependent on the environmental variables ("environmental models").
Initial inspection of the relationship between each biodiversity metric and temperature showed a hump-shaped pattern (Appendix S2), so we also included a quadratic term for temperature only.
Richness was fit with the negative binomial response family; others (SES of PD, SES of FD, SES of RPD, SES of RFD) were Gaussian.
We fit models including a Matérn correlation matrix based on the geographic centroids of each grid cell as a random effect with the R package `r pack("spaMM")` [@Rousset2014].
We verified the significance of each term of the model by conducting likelihood ratio tests (LRTs) between the full model and models each with a single term of interest removed, in sequence.
We also verified the significance of the full model relative to the null model (only including the Matérn correlation matrix) using LRTs.

We postulated that % apomictic taxa might explain the distribution of phylogenetic diversity better than temperature, so we also constructed a limited set of models to test this hypothesis ("reproductive models").
These models included biodiversity metrics with a phylogenetic component (SES of PD and SES of RPD) as dependent on % apomictic taxa, precipitation, and precipitation seasonality, with random spatial effects as described above.
We compared the fit between environmental models and reproductive models for SES of PD and SES of RPD using conditional Akaike information criterion (cAIC) [@Courtiol2017; @Vaida2005].

## Analysis of bioregions

We analyzed bioregions using the framework of @Daru2020 and their R package `r pack("phyloregion")`.
Briefly, the analysis involves first calculating beta diversity (change in composition between communities) using either species names (taxonomic bioregions) or phylogenetic distances (phylogenetic bioregions).
For taxonomic bioregions, we calculated the turnover in species between sites due to species replacement [@Baselga2012] with the Sørensen index.
For phylogenetic bioregions, we calculated phylogenetic dissimilarities based on the PhyloSor index [@Bryant2008].
The distances are used to construct a dendrogram, which is then split into into *k* bioregions.
The optimal value of *k* cannot be known a priori.
Previous vegetation zone schemes in Japan have typically included five or six zones [@Shimizu2014]; to enable approximate comparison with these, we tested values of *k* from 1 to 10, and selected the optimal value using the `r func("optimal_phyloregion")` function in the R package `r pack("phyloregion")`.

## Assessment of conservation status and threats

We downloaded SHP files of conservation zones in Japan from the Japan Ministry of the Environment (https://www.biodic.go.jp/biodiversity/activity/policy/map/map17/index.html) and Ministry of Land, Infrastructure, Transport and Tourism (https://nlftp.mlit.go.jp/ksj/gml/datalist/KsjTmplt-A45.html).
We excluded marine zones and those that do not protect plants.
We categorized protected areas as either high (no human activities allowed at all) or medium status [some economic activities allowed by permit\; @Kusumoto2017]; areas not afforded at least medium level of protection were not considered.
Conservation status in Japan is administered by multiple laws, and protected areas frequently overlap [@Natori2012].
To prevent double-counting, all protected areas within a protection status were merged, and protected areas that overlapped between medium and high status were considered only high status.

Besides habitat loss due to human activity, one major threat to ferns in Japan is herbivory by Japanese deer (*Cervus nippon*).
Although native, Japanese deer have caused extensive damage to native plant communities due to rapid population growth and range expansion since the 1970s [@Takatsuki2009].
Decrease or extirpation of fern populations due to deer herbivory has been observed in multiple sites and species across the country [@Minamitani2005; @Yahara2006; @Hattori2010].
Furthermore, 33 of 212 (15.5%) fern taxa listed as endangered (Red List status CR, EN, or VU) in Japan include herbivory by deer as one of the reasons for their endangered status [@MOJ2015; @kankyosho2022].
Although not all species of ferns are equally susceptible to deer herbivory [some are unpalatable to deer\; @Minamitani2005], these studies show it is an existential threat to many species.
Therefore, analysis of the threat posed to fern biodiversity by deer herbivory is needed to inform future conservation policies.

We downloaded distribution maps (SHP and TIF) of *Cervus nippon* from the Japan Ministry of the Environment (https://www.biodic.go.jp/biodiversity/activity/policy/map/map14/index.html).
To assess change in threat due to herbivory by deer over time, we used three maps available in this data set: range of Japanese deer surveyed in 1978, range surveyed in 2003, and estimated range inferred from a model including snow cover and forest type based on the 2003 survey data [@kankyosho2021].
We limited estimated range to areas considered "highly likely" to be occupied by deer [movement cost < 0.10\; @kankyosho2021].

We cropped grid cells to only include land area, then calculated the percentage of protected area of each status type or percentage of area occupied by deer for each survey period within grid cells with high levels of biodiversity as measured with PD, FD, PE, or taxonomic richness.
These values were compared with the overall percentage of protected area or area occupied by deer across Japan (baseline rate).
For PD, FD, and PE, significance was assessed with a one-tailed test at $\alpha$ = 0.05 against the 999 null communities described above.
For taxonomic richness, grid cells ranked in the top > 5% were considered highly diverse.

# RESULTS

## Datasets

```{r dataset-stats}
### Taxon list ###
# Make list of fern taxa `total_fern_taxa` in chunk `intro-stats`

### Phylogeny ###
phylogeny_sampled_percent <- ape::Ntip(japan_fern_tree) %>%
  magrittr::divide_by(total_fern_taxa) %>%
  scales::percent(accuracy = 0.01)

# Check height (age) of fern phylogeny
phylogeny_age <- japan_fern_tree %>%
  phytools::nodeHeights() %>%
  max() %>%
  round(1)

# Check monophyly of genera
mono_status_genera <-
  tibble(
    tip = japan_fern_tree$tip.label,
    tip_no = seq_along(tip),
    genus = str_split(tip, "_") %>% map_chr(1)
  ) %>%
  group_by(genus) %>%
  summarize(tip_vec = list(tip_no), .groups = "drop") %>%
  mutate(is_mono = map_lgl(
    tip_vec, ~ape::is.monophyletic(phy = japan_fern_tree, tips = .)))

# Assert only one genus is not monophyletic
assert_that(mono_status_genera %>% filter(is_mono == FALSE) %>% nrow == 1)

non_mono_genus <- mono_status_genera %>%
  filter(is_mono == FALSE) %>%
  pull(genus)

n_fern_phylo_genera <- mono_status_genera %>% pull(genus) %>% n_distinct()

# Check monophyly of families
mono_status_families <-
  tibble(
    tip = japan_fern_tree$tip.label,
    tip_no = seq_along(tip),
    genus = str_split(tip, "_") %>% map_chr(1)
  ) %>%
  left_join(ppgi, by = "genus") %>%
  group_by(family) %>%
  summarize(tip_vec = list(tip_no), .groups = "drop") %>%
  mutate(is_mono = map_lgl(
    tip_vec, ~ape::is.monophyletic(phy = japan_fern_tree, tips = .)))

# Assert all families are monophyletic
assert_that(mono_status_families %>% filter(is_mono == FALSE) %>% nrow == 0)

n_fern_phylo_families <- mono_status_families %>% pull(family) %>% n_distinct()

# Check monophyly of ferns
mono_status_class <-
  tibble(
    tip = japan_pterido_tree$tip.label,
    tip_no = seq_along(tip),
    genus = str_split(tip, "_") %>% map_chr(1)
  ) %>%
  left_join(ppgi, by = "genus") %>%
  group_by(class) %>%
  summarize(tip_vec = list(tip_no), .groups = "drop") %>%
  mutate(is_mono = map_lgl(
    tip_vec, ~ape::is.monophyletic(phy = japan_pterido_tree, tips = .)))

# Assert all classes (ferns, lycophytes) are monophyletic
assert_that(mono_status_class %>% filter(is_mono == FALSE) %>% nrow == 0)

### Specimens (filtered occurrence data) ###
n_specimens <- occ_sum$filtered$n_specimens$value %>%
  scales::number(big.mark = ",")
n_specimens_taxa <- occ_sum$filtered$n_taxa$value
n_specimens_taxa_percent <- n_specimens_taxa %>%
  magrittr::divide_by(total_fern_taxa) %>%
  scales::percent(accuracy = 0.01)

# Calculate mean redundancy
mean_redundancy <-
  biodiv_ferns_spatial %>%
  select(redundancy) %>%
  st_set_geometry(NULL) %>%
  summarize(
    mean = mean(redundancy),
    sd = sd(redundancy)
  ) %>%
  mutate(text = glue::glue("{round(mean, 2)} ± {round(sd, 2)}")) %>%
  pull(text)

# Calculate number of excluded cells due to low redundancy
n_excluded_cells <- nrow(comm_ferns_full) - nrow(comm_ferns)

### Traits ###

# Count number of traits of each type in final dataset
n_binary_traits <- traits_summary %>%
  filter(trait_type == "binary") %>%
  nrow() %>%
  jntools::english2()
n_cont_traits <- traits_summary %>%
  filter(trait_type == "continuous") %>%
  nrow() %>%
  jntools::english2()
n_qual_traits <- traits_summary %>%
  filter(trait_type == "qualitative") %>%
  nrow() %>%
  jntools::english2()
n_qual_traits_as_binary <- traits_summary %>%
  filter(trait_type == "qualitative") %>%
  pull(n) %>%
  sum() %>%
  jntools::english2()

# Calculate number/percent of missing data in trait matrix
trait_sampling_summary <-
fern_list_non_hybrid %>%
  select(taxon) %>%
  verify(all(fern_traits$taxon %in% taxon)) %>%
  left_join(
    transmute(fern_traits, taxon, sampled = TRUE), by = "taxon"
  ) %>%
  mutate(sampled = replace_na(sampled, FALSE))

n_taxa_traits_sampled <- sum(trait_sampling_summary$sampled)

percent_taxa_traits_sampled <-
  n_taxa_traits_sampled %>%
  magrittr::divide_by(total_fern_taxa) %>%
  scales::percent(accuracy = 0.1)

fern_traits_missing_summary <-
  fern_traits %>%
  pivot_longer(-taxon) %>%
  mutate(missing_data = is.na(value)) %>%
  summarize(
    total = n(),
    missing = sum(missing_data)
  )

percent_traits_missing <-
  fern_traits_missing_summary$missing %>%
  divide_by(fern_traits_missing_summary$total) %>%
  scales::percent(accuracy = 0.01)
```

The phylogeny included `r ape::Ntip(japan_fern_tree)` taxa (`r phylogeny_sampled_percent` of the total number of native, nonhybrid ferns [`r total_fern_taxa` taxa]) representing `r n_fern_phylo_genera` genera and `r n_fern_phylo_families` families.
The topology generally agreed with earlier, large plastid fern phylogenies [@Schuettpelz2007; @Lehtonen2011; @Testo2016a\; `r s_figure("tree")`].
Ferns were recovered as monophyletic, as were all families and genera except for *`r non_mono_genus`* [families follow @PteridophytePhylogenyGroupI2016].
The age of the fern clade is estimated to be `r phylogeny_age` Ma.

The cleaned list of specimens included `r n_specimens` specimens representing `r n_specimens_taxa` taxa (`r n_specimens_taxa_percent`).
The distribution of collection locations was highly skewed, with > 3000 specimens per grid cell northwest of Tokyo in Saitama Prefecture, and many fewer elsewhere (`r s_figure("abundance")`A).
Despite this skew, sampling redundancy was generally good [> 0.3\; @Gonzalez-Orozco2014b] throughout the country (mean `r mean_redundancy`, final community data matrix; SDs are given unless mentioned otherwise; `r s_figure("abundance")`B).
The final community data matrix included `r ncol(comm_ferns)` taxa and `r nrow(comm_ferns) %>% scales::number(big.mark = ",")` grid cells. `r n_excluded_cells` grid cells were excluded due to low redundancy (< 0.1); these were mostly from Hokkaido and coastal areas.

After removing correlated and low-variance traits, the final trait set included `r n_binary_traits` binary trait, `r n_cont_traits` continuous traits, and `r n_qual_traits` qualitative traits (coded as `r n_qual_traits_as_binary` binary traits; Tables `r table_num("traits-used")`, `r s_table_num("phy-sig-binary")`).
The trait matrix included `r n_taxa_traits_sampled` taxa (`r percent_taxa_traits_sampled`) and `r fern_traits %>% select(-taxon) %>% ncol()` traits, with `r percent_traits_missing` missing data.

## Phylogenetic signal

```{r phy-sig-stats}
# Verify that min lambda >0.95
assert_that(phy_sig_results %>% pull(lambda) %>% min > 0.95)
# Verify that max K < 0.6
assert_that(phy_sig_results %>% pull(kval) %>% max < 0.10)

# Sum the number of traits with D <= 0.5 and not random
n_d_signif <-
  binary_sig_results %>%
  filter(prob_random < 0.05, D <= 0.5) %>%
  pull(trait) %>%
  n_distinct()
```

All continuous traits showed significant phylogenetic signal, but estimates of signal strength differed between *K* and $\lambda$: all $\lambda$ values were > 0.95, indicating evolution by BM; however, values of *K* were < 0.1, indicating less phylogenetic signal than expected under BM (`r s_table("phy-sig")`).
`r n_d_signif` out of `r nrow(binary_sig_results)` of binary traits (including qualitative traits coded as binary) showed evidence of phylogenetic signal (*D* ≤ 0.5, significantly different from random; `r s_table("phy-sig-binary")`).

## Observed diversity

Taxonomic richness is lowest on the northernmost island of Hokkaido, then increases with decreasing latitude until reaching a maximum in southern Kyushu, then decreases again in smaller islands further south (`r figure("raw-diversity")`A, `r s_figure("raw-lat")`A).
Observed values of PD and FD were highly correlated with richness as expected (`r figure("raw-diversity")`B, C), showing an initial steep slope that gradually starts to level off at higher richness (`r s_figure("raw-biplot")`A, C).
Observed values of RPD and RFD were not linearly related to richness; however, they showed high variance at low species richness, and low variance at high species richness (`r s_figure("raw-biplot")`B, D).

## Randomization tests for phylogenetic and functional diversity

Grid cells with significantly low PD were almost all located on the main islands of Japan (Honshu, Kyushu, and Shikoku; `r figure("rand-diversity")`A).
A smaller number of grid cells had significantly high PD, and these were located mostly on small southern (Ryukyu, Izu, Ogasawara) islands or coastal areas (`r eg`, Kii Peninsula).
Grid cells with significantly low RPD were mostly observed in the southern main islands, particularly in and around Kyushu (`r figure("rand-diversity")`B).
Grid cells with significantly high RPD are located both in the small southern islands and northern Honshu and Hokkaido.
Results of the randomization test for functional diversity were broadly similar to phylogenetic diversity (`r figure("rand-diversity")`C, D).

## Relationships between biodiversity and climate or reproductive mode

```{r discus-model-stats}
# Check that conditional AIC was lower in the reproductive model compared to
# the environmental model for SES of RPD
# Split aic_env_repro dataframe into nested list for easier indexing
c_aic <- aic_env_repro %>% split(.$resp_var) %>% map(~split(., .$model_type))
assert_that(
  c_aic$rpd_obs_z$percent_apo$conditional_aic <
    c_aic$rpd_obs_z$temp$conditional_aic)
# But not for SES of PD
assert_that(
  c_aic$pd_obs_z$percent_apo$conditional_aic >
    c_aic$pd_obs_z$temp$conditional_aic)
```

All models were significant relative to null models without any fixed effects (`r s_table("lrt-null")`).
The effect of temperature dominated (had the greatest absolute *t*-value) in environmental models (`r figure("coeff")`A--E), and % apomictic taxa dominated in reproductive models (`r figure("coeff")`F--G); temperature and % apomictic taxa were significant in each model where they were included, whereas other predictors (area, precipitation, precipitation seasonality) were not significant in some models.
Richness showed a hump-shaped relationship with temperature, reaching maximum richness at intermediate temperatures (`r figure("modelfit")`A).
Phylogenetic and functional diversity showed either moderately (SES of PD, SES of FD) to strongly (SES of RPD, SES of RFD) hump-shaped relationships with temperature (`r figure("modelfit")`B--E).
SES of RPD showed a clear, linear decline with % apomictic taxa (`r figure("modelfit")`G); a similar pattern was detected in SES of PD, but with wider confidence intervals (`r figure("modelfit")`F).
Conditional AIC was lower (indicating better model fit) in the reproductive model compared to the environmental model for SES of RPD, but not SES of PD (`r s_table("aic")`).

## Categorical analysis of neo- and paleo-endemism

Nearly all grid cells in islands south of Kyushu (Ryukyu and Ogasawara) have significant levels of phylogenetic endemism, with low (non-significant) endemism observed elsewhere (`r figure("endemism")`).
The vast majority of grid cells with significant endemism are mixed or super-endemic.
Concentrations of paleoendemism were observed in the Ryukyu archipelago, on Okinawa, Miyako, and Iriomote Islands.
Small clusters of grid cells containing significant endemism were also observed northwest of Tokyo (Nagano prefecture), and in Hokkaido (`r figure("endemism")`).
When we ran CANAPE with a reduced data set including only taxa absolutely restricted to Japan to account for the border effect, there were fewer significant grid cells, but a similar pattern of significant endemism mostly occurring in the southern islands was observed (`r s_figure("endemism-restricted")`).

## Bioregions

```{r verify-cutoff}
assert_that(bioregion_cutoff == 2, msg = "Bioregions cutoff for lumping not 2!")
```

Clustering analysis grouped grid cells into `r english2(k_taxonomy$optimal$k)` or `r english2(k_phylogeny$optimal$k)` bioregions (`r s_figure("k-plot")`) on the basis of taxonomic or phylogenetic distances, respectively.
The vast majority of grid cells fell into a subset of these bioregions: using taxonomic distances, they include bioregion 1 (Hokkaido and high elevation areas of northern Honshu), bioregion 2 (low elevation areas of northern Honshu, southern Honshu, Kyushu, and Shikoku), bioregion 3 (Ryukyu Islands), and bioregion 4 (Ogasawara Islands) (`r figure("bioregions")`A; `r s_figure("dendrogram")`A).
Phylogenetic distances produced similar results, but the Ryukyu and Ogasawara Islands merged into a single bioregion (3), and bioregion 1 covered most of northern Honshu, while extending further south (`r figure("bioregions")`B; `r s_figure("dendrogram")`B).
The other much smaller bioregions (including only one or two grid cells each) are likely artifacts of insufficiently low taxon richness for robust clustering and are not discussed further.

Taxonomic bioregions 3 and 4 have higher mean SES values of PD, RPD, FD, and RFD than bioregions 1 and 2 (`r figure("div-by-tax-region")`).
Bioregions 3 and 4 also have consistently high PE *p*-scores, whereas bioregions 1 and 2 have a much wider variation (`r figure("div-by-tax-region")`E).
Bioregion 2 has lower RPD and RFD than bioregion 1.
Comparison of biodiversity metrics across phylogenetic bioregions was similar, except that phylogenetic bioregion 3 mostly corresponds to taxonomic bioregions 3 and 4 combined (`r s_figure("div-by-phy-region")`).

## Conservation status and threats

```{r conserv-stats}
# Convert signif_cells_protected_area to wide format
signif_cells_protected_area_wide <- #nolint
  signif_cells_protected_area %>%
  select(status, metric, percent_protected) %>%
  pivot_wider(names_from = status, values_from = percent_protected) %>%
  rowwise %>%
  mutate(total = sum(medium, high)) %>%
  ungroup

# Calculate min and max percent area protected overall
# - define helper functions
calc_min_pa <- function(signif_cells_protected_area_wide, ...) { # nolint
  signif_cells_protected_area_wide %>%
    pull(...) %>%
    min() %>%
    scales::percent(accuracy = 0.1)
}
calc_max_pa <- function(signif_cells_protected_area_wide, ...) { # nolint
  signif_cells_protected_area_wide %>%
    pull(...) %>%
    max() %>%
    scales::percent(accuracy = 0.1)
}

min_protected_med <- calc_min_pa(signif_cells_protected_area_wide, medium)
max_protected_med <- calc_max_pa(signif_cells_protected_area_wide, medium)
min_protected_high <- calc_min_pa(signif_cells_protected_area_wide, high)
max_protected_high <- calc_max_pa(signif_cells_protected_area_wide, high)
min_protected_total <- calc_min_pa(signif_cells_protected_area_wide, total)
max_protected_total <- calc_max_pa(signif_cells_protected_area_wide, total)

# Calculate min and max protected area for the best protected category
signif_cells_protected_area_best <- # nolint
  signif_cells_protected_area_wide %>%
  slice_max(total)

# Verify that the best-protected biodiv type is PE
assertthat::assert_that(signif_cells_protected_area_best$metric == "pe")

best_protected_med <- signif_cells_protected_area_best %>%
  pull(medium) %>%
  scales::percent(accuracy = 0.1)
best_protected_high <- signif_cells_protected_area_best %>%
  pull(high) %>%
  scales::percent(accuracy = 0.1)
best_protected_total <- signif_cells_protected_area_best %>%
  pull(total) %>%
  scales::percent(accuracy = 0.1)

# Calculate the total area of Japan
area_of_japan <-
  japan_shp %>%
  mutate(
    area = st_area(.) %>%
      units::set_units(km^2)
  ) %>%
  pull(area)

# Calculate baseline protection percentage:
# percent protected areas across the whole country
total_percent_protected <-
  protected_areas %>%
  st_make_valid %>%
  mutate(
    area = st_area(.) %>%
      units::set_units(km^2)
  ) %>%
  st_drop_geometry %>%
  as_tibble() %>%
  filter(status %in% c("medium", "high")) %>%
  group_by(status) %>%
  summarize(
    total_area = sum(area),
    .groups = "drop"
  ) %>%
  mutate(percent_protected = total_area / area_of_japan) %>%
  mutate(percent_protected = as.numeric(percent_protected)) %>%
  # Add sum for high + medium
  bind_rows(
    summarize(
      .,
      percent_protected = sum(percent_protected),
      total_area = sum(total_area)
    ) %>% mutate(status = "total")
  )

# Extract values for medium and high protection level
total_percent_protected_medium <- total_percent_protected %>%
  filter(status == "medium") %>%
  pull(percent_protected)
total_percent_protected_high <- total_percent_protected %>%
  filter(status == "high") %>%
  pull(percent_protected)
total_percent_protected_all <- sum(
  total_percent_protected_medium, total_percent_protected_high)

### Deer ###

# Calculate baseline protection percentage:
# percent protected areas across the whole country
total_percent_deer_danger <-
  deer_range %>%
  mutate(
    area = st_area(.) %>% units::set_units(km^2)
  ) %>%
  st_drop_geometry %>%
  as_tibble() %>%
  group_by(range) %>%
  summarize(
    total_area = sum(area),
    .groups = "drop"
  ) %>%
  mutate(percent_danger = total_area / area_of_japan) %>%
  mutate(percent_danger = as.numeric(percent_danger))

# Convert to a list for easy indexing
deer_overall <- total_percent_deer_danger %>%
  mutate(
    # Format numbers for nice printing
    area = total_area %>%
      as.numeric %>%
      scales::number(accuracy = 1, big.mark = ",") %>%
      paste("km^2^"),
    percent = scales::percent(percent_danger, accuracy = 0.1)
  ) %>%
  split(.$range)

deer_in_high_biodiv <-
signif_cells_deer_danger %>%
  group_by(range) %>%
  summarize(percent_danger = mean(percent_danger), .groups = "drop") %>%
  mutate(
    # Format numbers for nice printing
    percent = scales::percent(percent_danger, accuracy = 0.1)) %>%
  split(.$range)

deer_highest_danger <- signif_cells_deer_danger %>%
  slice_max(percent_danger) %>%
  # Make sure biodiv type with greatest danger is richness in estimated dataset
  verify(metric == "richness") %>%
  verify(range == "estimated") %>%
  mutate(
    # Format numbers for nice printing
    percent = scales::percent(percent_danger, accuracy = 0.1)) %>%
  split(.$range)
```

The percentage of protected area in grid cells with significantly high biodiversity range from `r min_protected_total` to `r max_protected_total` total, including `r min_protected_high` to `r max_protected_high` with high status and `r min_protected_med` to `r max_protected_med` with medium status (`r figure("conserv-status")`A).
Cells with significantly high biodiversity had a similar or greater percentage of area protected compared to the baseline rate for Japan at the medium protection status (`r scales::percent(total_percent_protected_medium, accuracy = 0.1)`) across all biodiversity metrics.
In contrast, grid cells with significantly high biodiversity had a similar or lower percentage of area protected than the baseline rate for Japan at the high protection status (`r scales::percent(total_percent_protected_high, accuracy = 0.1)`) for most biodiversity metrics.
Only grid cells with significantly high PE had a higher percentage area protected (`r best_protected_high` high, `r best_protected_med` medium, `r best_protected_total` total) than the baseline rate at both levels of protection.

The range of Japanese deer in Japan increased from `r deer_overall[["1978"]]$area` (`r deer_overall[["1978"]]$percent`) in 1978 to `r deer_overall[["2003"]]$area` (`r deer_overall[["2003"]]$percent`) in 2003, and may be as high as `r deer_overall$estimated$area` (`r deer_overall$estimated$percent`) according to model estimates.
The percentage of area occupied by deer in grid cells with significantly high biodiversity increased from `r deer_in_high_biodiv[["1978"]]$percent` in 1978 to `r deer_in_high_biodiv[["2003"]]$percent` in 2003 when averaged across biodiversity types.
Grid cells with high PD, FD, or PE tended to have less deer occupation compared to the baseline rate (except for PE relative to 1978 levels; `r figure("conserv-status")`B).
However, grid cells with high species richness consistently exceeded the baseline rate of deer presence, with up to `r deer_highest_danger$estimated$percent` occupied by deer in the estimated range data set.

# DISCUSSION

## Distribution and drivers of biodiversity

```{r distribution-stats}
# Summarize maximum latitude by genus, family
max_lat_summary <- summarize_fern_lat_max(lat_span_summary, ppgi)

# Calculate number of genera and families restricted to lat < 30.1
southern_genera <-
  max_lat_summary %>%
  filter(max_lat < 30.1, taxon_level == "genus")

n_southern_genera <- southern_genera %>%
  pull(higher_taxon) %>%
  n_distinct() %>%
  english2()

southern_families <-
  max_lat_summary %>%
  filter(max_lat < 30.1, taxon_level == "family")

n_southern_families <- southern_families %>%
  pull(higher_taxon) %>%
  n_distinct() %>%
  english2()
```

The latitudinal hump-shaped pattern of species richness observed in ferns here (Figures `r figure_num("raw-diversity")`A, `r s_figure_num("raw-lat")`A) is well-documented [@Kusumoto2017; @Ebihara2019b].
Similar patterns have been observed in deciduous broadleaved trees and conifers [@Shiono2015], and niche-modeled distributions of woody plants in Japan [@Fukaya2020].
Our spatial modeling analysis indicates that this is strongly linked to temperature, with maximum richness occurring at intermediate temperatures (`r figure("modelfit")`A).
Previous studies have identified temperature, water availability [@Bickford2006; @Qian2012], or a combination of these [@Nagalingum2015; @Link-Perez2018; @Qian2021] to be important drivers of spatial fern diversity.
A hump-shaped pattern of richness along elevation (maximum richness at intermediate elevations) has been frequently observed in ferns on mountains in tropical or subtropical areas [@Kessler2011a; @Suissa2021].
Hypotheses to explain such a pattern include purely geometrical constraints as well as biological hypotheses positing maximally favorable environmental conditions at mid-elevations [@Colwell2004a].
Another reason for the greater richness on the main islands compared to southern islands in Japan is that the main islands harbor higher elevational variation that may support a wider range of taxa than is present in the southern islands.

The overwhelming pattern characterizing phylogenetic and functional diversity (PD and FD) is the high diversity of southern subtropical islands (Ryukyu and Ogasawara Isl.) (`r figure("rand-diversity")`A, C).
This is almost certainly due to the presence of primarily tropical lineages that do not occur in other parts of the country, which lack subtropical climate; `r n_southern_genera` genera and `r n_southern_families` families only occur south of 30.1° latitude (just south of Yakushima Isl.).
The similarity in patterns of PD and FD is likely due to the moderate degree of phylogenetic signal present in the traits used in our analysis (Tables `r s_table_num("phy-sig-binary")`, `r s_table_num("phy-sig")`), and suggests that at least in the ferns of Japan, phylogeny can be used as a reasonable stand-in for functional diversity.
Given their high PD, it is perhaps not surprising that southern islands also host high amounts of phylogenetic endemism (PE); indeed, the vast majority of grid cells with significant levels of PE, as well as paleoendemic grid cells, are located in the southern islands (Figures `r figure_num("endemism")`, `r s_figure_num("endemism-restricted")`).
This is clearly due to the small land area of these islands: small area combined with high PD is expected to lead to high PE.

Another striking pattern revealed here is the predominance of long branches in the northern half of Honshu and the southern islands, and the predominance of short branches in southern Honshu, Shikoku, and Kyushu (`r figure("rand-diversity")`B).
The long branches in the southern islands are likely being driven by the tropical lineages restricted to this area; those in the north may be due to refugial lineages that are distantly related to others and become more species rich at high latitudes, such as *Equisetum* (`r s_figure("genus-map")`A). The preponderance of short branches in southern Honshu, Shikoku, and Kyushu may be due to the radiation of certain species-rich genera there such as *Deparia* (`r s_figure("genus-map")`B) and *Dryopteris* (`r s_figure("genus-map")`D).

Furthermore, we identified % apomictic taxa as a strong driver of SES of RPD (Figures `r figure_num("coeff")`G, `r figure_num("modelfit")`G), and to a slightly lesser extent SES of PD (Figures `r figure_num("coeff")`F, `r figure_num("modelfit")`F).
Indeed, *Dryopteris* is one such genus with a high rate of apomictic taxa (`r s_figure("apo-genus")`), and other genera with high rates of apomixis (*Cyrtomium*, *Pteris*; `r s_figure("apo-genus")`) also reach greatest richness in the southern main islands (`r s_figure("genus-map")`).
This is consistent with the hypothesis that communities with high proportions of asexual taxa have lower PD, and is the first time to our knowledge that this has been demonstrated in ferns.
We verified that this relationship is also observed when analyzed with a phylogeny with branch lengths in units of genetic change (`r ie`, the output of the maximum likelihood phylogenetic analysis, before dating), a more direct measure of genetic diversity (`r s_figure("modelfit-not-ult")`).
While temperature is also correlated with % apomictic taxa and therefore cannot be discounted, conditional AIC indicates a better fit of the reproductive model than the environmental model for SES of RPD (`r s_table("aic")`).
Ultimately, it is likely that temperature, in combination with other environmental conditions, is driving the predominance of apomictic taxa: @Tanaka2014 also found that the proportion of apomictic taxa reaches a maximum at ca. 34.8 °N latitude and increases with temperature in Japan (their study was limited to the main islands and did not include the smaller southern islands).
@Tanaka2014 asserted that one limiting factor affecting establishment of apomictic ferns at high latitudes or cold climates is the large cell size of apomicts, which are typically polyploid.
As fern gametophytes are only a single cell-layer in thickness with no cuticle or stomata, it is expected they would be particularly susceptible to freezing damage induced by larger cells with higher water content; however, no experimental studies have tested these hypotheses to our knowledge.

Water limitation is thought to play an important role in promoting apomixis in ferns especially in deserts [@Grusz2021], since ferns depend on water for transfer of sperm to egg, and apomictic plants would be able to reproduce in the absence of water.
A similar argument has been made to assert that areas with seasonal monsoons are linked to increased rates of apomixis in ferns [@Liu2012; @Tanaka2014; @Picard2021], but in Japan, although the areas including high percentages of apomictic taxa are monsoonal, the degree of seasonal water limitation does not approach that of a desert.
The monsoon hypothesis was not supported by our results, which did not show a particularly strong effect of precipitation seasonality in most models (`r figure("coeff")`), nor a correlation with % apomictic taxa (`r s_table("corr")`).

The variation in distribution of different biodiversity metrics is reflected by the bioregions analysis, which tended to group grid cells with similar biodiversity values (`r figure("div-by-tax-region")`).
Interestingly, the bioregions identified here correspond roughly to major vegetation zones previously categorized on an *ad-hoc* basis; bioregion 1 corresponds to the cool or subartic deciduous type, bioregion 2 to the warm evergreen type, and bioregions 3 and 4 to the subtropical rainy type [@Shimizu2014].
Furthermore, the border between bioregions 1 and 2 approximately corresponds to the "*Gleichenia japonica* Line" proposed by @Nakaike1983, which he proposed based on the distribution of *Diplopterygium glaucum* (Houtt.) Nakai (= *Gleichenia japonica* Spreng.) as an indicator of evergreen broad-leaf forests in Japan.
This suggests that ferns are useful as bioindicators in Japan.
Although some studies have emphasized the importance of deep-sea straits (the Tsugaru Strait separating Hokkaido from Honshu and the Tokara Strait separating the southern islands from Kyushu) in structuring biological communities in Japan [@Millien-Parra1999; @Kubota2014], these did not play a major role in structuring fern bioregions.
The Tsugaru Strait had no relation to any of the bioregions, and although the Tokara Strait somewhat splits phylogenetic bioregions 2 and 3, there is some overlap between them on Yakushima and Amami Isl. (`r figure("bioregions")`B).
The weak effect of these straits as barriers makes sense given the presumably high dispersal abilities of ferns, which disperse via tiny spores carried on the wind [@Tryon1970].
Rather, fern bioregions in Japan broadly separate along gradients of temperature and precipitation; in particular, precipitation splits taxonomic bioregion 3 (mostly the Ryukyu islands; wetter) from bioregion 4 (mostly the Ogasawara islands; drier) (`r s_figure("bioregions-env")`).

## Conservation status and threats

We found that grid cells with significantly high amounts of biodiversity are similarly or better protected compared to the baseline protection rate at the medium protection level across all biodiversity metrics, and better than the baseline rate at the high protection level for PE (`r figure("conserv-status")`A).
The high protection status of grid cells with high PE may be due to an emphasis on conserving endemic plant and animal species in Japanese southern islands such as Iriomote and Amami [@kankyosho2021b], which include many grid cells with high PE (`r s_figure("protected-map")`D).
One possible reason for the generally high level of protection overall is that in the past, some ferns have been included in lists of rare and endemic plants for consideration in designation of conservation areas [A. Ebihara, pers. comm.\; @kankyosho2021b].
Also, (as suggested previously) ferns may serve as bioindicators reflecting overall levels of biodiversity, particularly for other plant groups.

Despite the good news that ferns enjoy relatively strong protection status in Japan, we must bear in mind that there are other threats posed that conservation zones cannot ameliorate, the most obvious being climate change.
One other such threat is herbivory by native Japanese deer, *Cervus nippon* [@Minamitani2005; @Yahara2006; @Hattori2010; @kankyosho2022].
The rapid increase in population size observed in surveys from 1978 to 2003 is likely due to a combination of climate change and a decrease in human hunters and natural predators [@Takatsuki2009].
At dense populations, the deer can prevent forest regeneration and result in plant communities dominated by a limited number of species unpalatable to deer [@Takatsuki2009], although lower population sizes may promote plant diversity as per the intermediate disturbance hypothesis [@Nishizawa2016].
They have caused extensive damage even within national parks [@Tokita2006], which have the highest protection status.
Although various efforts to prevent deer herbivory have been deployed such as culling and exclusion fences [@Takatsuki2009], much work remains to be done to save vulnerable plant populations.

Our analysis shows that the threat to fern biodiversity posed by Japanese deer differs strongly by biodiversity type; although grid cells with significantly high FD, PD, and PE tend to have lower rates of deer occupancy compared to the baseline rate in recent surveys or modeled data, grid cells with high richness experience much greater co-occurrence with deer than the baseline rate (`r figure("conserv-status")`B).
This pattern is due to the fact that Japanese deer do not occur in the southern islands, which house high amounts of PD and PE; rather, areas with high species richness are primarily located in southern Honshu, Shikoku, and Kyushu, where deer expansion has been intense (`r s_figure("deer-map")`A).
This result highlights the importance of using multiple metrics of biodiversity for establishing conservation priorities.

# CONCLUSIONS

```{r limits-stats}
# Find the species with the widest latitudinal distribution
taxon_with_widest_lat <- lat_span_summary %>%
  mutate(lat_span = max_lat - min_lat) %>%
  slice_max(lat_span) %>%
  # Add formatted scientific name
  left_join(
    tibble(
      taxon = "Pteridium_aquilinum_japonicum",
      text = "*Pteridium aquilinum* (L.) Kuhn subsp. *japonicum* (Nakai) A. et S. Löve" #nolint
    ),
    by = "taxon"
  ) %>%
  # Round off latitude
  mutate(across(c(min_lat, max_lat), ~round(., 2)))

# Make sure the taxon_with_widest_lat is what we expect
assert_that(
  taxon_with_widest_lat$taxon == "Pteridium_aquilinum_japonicum",
  msg = "Taxon with widest latitude wasn't Pteridium_aquilinum_japonicum")

# Find number of species occurring across areas defined by straits
n_widespread_taxa <-
  lat_span_summary %>%
  # Filter by approx location of Tokara strait (29) and Tsugaru Strait (41.5)
  filter(min_lat < 29, max_lat > 41.5) %>%
  pull(taxon) %>%
  n_distinct %>%
  english2

# Check number of hybrid fern taxa
n_hybrid_fern_taxa <- fern_list %>%
  filter(hybrid == TRUE) %>%
  pull(taxon) %>%
  n_distinct()
```

Our study integrates diverse, thoroughly collected datasets to characterize the spatial biodiversity of Japanese ferns and gain insight into the processes generating these patterns.
We conclude by summarizing some caveats that must be kept in mind when interpreting our results, while also suggesting avenues for future research.

First, the data include specimens collected over a wide time period, so they do not necessarily represent the current distribution of ferns in Japan.
Furthermore, the collections are not evenly collected throughout the country, but rather some areas have many more specimens than others (`r s_figure("abundance")`).
Indeed, one promising avenue of research may be to take advantage of the most densely sampled areas to investigate changes in abundance over time.

The extreme evolutionary distinctiveness of the southern islands is both a major feature of the Japanese fern flora and a challenge for analysis.
The tropical lineages occurring only in these islands are likely responsible for the comparatively low (`r ie`, clustered) values of PD observed on the main islands (`r figure("rand-diversity")`A).
One way to account for this bimodal distribution pattern would be to use a restricted null model that does not allow all taxa to disperse anywhere within the study area [@Mishler2020].
However, ferns are known to have high dispersal abilities [@Tryon1970; @Barrington1993; @Moran2001].
In our data set, the taxon with the greatest latitudinal distribution (`r taxon_with_widest_lat$text`) spans nearly the entire region (`r taxon_with_widest_lat$min_lat`°N to `r taxon_with_widest_lat$max_lat`°N), and `r n_widespread_taxa` taxa have latitudinal distributions spanning the Tokara strait in the south (29°N) to the Tsugaru strait in the north (41.5°N).
Therefore, it seems reasonable to assume that ferns are capable of dispersing across the entire study area.
Furthermore, it should be noted that the southern islands are all low-altitude (not exceeding 1,000 m).
If high elevation habitats were available, the southern islands would be expected to share more species in common with areas further north.
For example, Taiwan, which is just east of the southernmost Japanese islands and harbors high mountains (> 3,000 m), shares many fern species with the main Japanese islands that are not found in southern Japanese islands.
This supports that the sharp taxonomic turnover between southern and main islands is more likely due to ecological factors than dispersal limitation.

One element of biodiversity that we did not consider here is hybrids.
The Japanese fern flora includes `r n_hybrid_fern_taxa` nothotaxa [@Ebihara2019b].
Although hybrids are often thought of as evolutionary "dead-ends", in some cases they are able to interbreed with diploids and thus contribute to diversification [@Barrington1989].
Future studies should consider the distribution of hybrids in Japan and their possible effects on biodiversity.
Methods for calculating PD would need to be modified to take into account multiple lineages contributing to one terminal taxon.

Finally, a long under-appreciated aspect of fern ecology is the role of the gametophyte.
Unlike seed plants, ferns have gametophytes that are capable of growing independently from the sporophyte.
Futhermore, the two stages of the life cycle have vastly different morphology and ecophysiology, and may even occur over partially to completely disjunct ranges [reviewed in @Pinson2016a].
Here, as in the vast majority of fern ecology studies, we consider only the sporophyte stage.
However, recent studies in Japan [@Ebihara2013] and elsewhere [@Nitta2017] have revealed different patterns in the community structure of gametophytes and sporophytes.
The comprehensive molecular sampling of sporophytes in Japan makes this area ideal for conducting high-throughput DNA sequencing analyses to compare patterns of biodiversity between life stages in ferns across large spatial scales.

## ACKNOWLEDGEMENTS

The authors thank Members of the Iwasaki Lab at the University of Tokyo and two anonymous reviewers for providing comments that improved the manuscript.
Shawn Laffan provided advice about CANAPE.
This study was supported by JSPS KAKENHI Grant Number 16H06279.
The authors are grateful to members of the Nippon Fernist Club for their cooperation and efforts towards quantifying and conserving the biodiversity of the ferns of Japan.

## AUTHOR CONTRIBUTIONS

J.H.N., B.M., and A.E. conceived the ideas; A.E. provided the data; J.H.N. analyzed the data; W.I provided resources; J.H.N. wrote the manuscript with input from all co-authors.

## DATA AVAILABILITY STATEMENT

Code to replicate all analyses, figures, and this manuscript are available at https://github.com/joelnitta/japan_ferns_spatial_phy.
A Docker image to run the code is available at https://hub.docker.com/r/joelnitta/japan_ferns_spatial_phy.
Data files needed to run the analysis and selected results files are available from Figshare at http://doi.org/10.6084/m9.figshare.16655263.

## SUPPORTING INFORMATION

Additional supporting information may be found online in the Supporting Information section at the end of the article.

**Appendix S1**. Supplementary tables and figures, including Tables `r s_table_num("moran")`--`r s_table_num("aic")` and Figures `r s_figure_num("sampling-curve")`--`r s_figure_num("deer-map")`.

**`r s_table("moran")`**. Spatial autocorrelation in biodiversity metrics (richness, SES of PD, SES of RPD, SES of FD, SES of RFD) and independent variables (environmental variables and % apomictic taxa) as measured with Moran's *I*.  
**`r s_table("corr")`**. Results of modified *t*-test for correlation between variables while taking into account spatial position.  
**`r s_table("phy-sig-binary")`**. Phylogenetic signal in quantitative (binary) functional traits of the ferns of Japan.  
**`r s_table("phy-sig")`**. Phylogenetic signal in continuous functional traits of the ferns of Japan.  
**`r s_table("lrt-null")`**. Likelihood ratio test (LRT) between full model and null model (model only including the spatial Matérn correlation matrix).  
**`r s_table("aic")`**. Model fit as measured with conditional Akaike Information Criterion (cAIC).  
**`r s_figure("sampling-curve")`**. Species collection curve for the ferns of Japan.  
**`r s_figure("grain-size")`**. Effect of grain size on sampling redundancy.  
**`r s_figure("envir")`**. Reproductive mode and environmental data in 20 km grid cells across Japan.  
**`r s_figure("tree")`**. Time tree of native Japanese ferns and lycophytes, excluding nothotaxa.  
**`r s_figure("abundance")`**. Observed number of specimens and sampling redundancy per 20 km grid cell in the ferns of Japan.  
**`r s_figure("raw-lat")`**. Raw taxonomic diversity, phylogenetic diversity, and functional diversity of the ferns of Japan plotted by latitude.  
**`r s_figure("raw-biplot")`**. Relationships between observed functional and phylogenetic diversity and taxonomic richness in the ferns of Japan.  
**`r s_figure("endemism-restricted")`**. Phylogenetic endemism of the ferns of Japan measured using CANAPE (categorical analysis of neo- and paleo-endemism), restricted data set including only taxa endemic to Japan.  
**`r s_figure("k-plot")`**. Selection of *k* for bioregions.  
**`r s_figure("dendrogram")`**. Bioregion dendrograms.  
**`r s_figure("div-by-phy-region")`**. Phylogenetic and morphological diversity of the ferns of Japan by phylogenetic bioregion.    
**`r s_figure("genus-map")`**. Map of species richness for selected genera of ferns of Japan.  
**`r s_figure("apo-genus")`**. Rate of apomixis in genera of Japanese ferns with > 1 apomictic taxon.  
**`r s_figure("modelfit-not-ult")`**. Relationship between phylogenetic diversity (PD) and relative PD and % apomictic taxa, models inferred using phylogeny with branchlengths in units of expected genetic change (untransformed ML tree).  
**`r s_figure("bioregions-env")`**. Scatterplots of grid cell membership in taxonomic or phylogenetic bioregions arranged by mean annual temperature and annual precipitation.  
**`r s_figure("protected-map")`**. Maps showing overlap of grid cells with significantly high biodiversity for the ferns of Japan and protected areas.  
**`r s_figure("deer-map")`**. Maps showing overlap of grid cells with significantly high biodiversity for the ferns of Japan and distribution of Japanese deer (*Cervus nippon*).  

**Appendix S2**. Exploratory data analysis for spatial modeling.

## References
<!-- Insert <div id="refs"></div> to place the refs here instead of at the end -->
<div id="refs"></div>

\newpage

| Trait               | Significance                                               | Reference       |
|---------------------|------------------------------------------------------------|-----------------|
| Indusium (b)        | Protects developing spores                                 | @Poppinga2015   |
| Frond width (c)     | Scales with overall plant size                             | @Creese2011     |
| No. pinna pairs (c) | Less divided fronds have lower rates of evapotranspiration | @Kluge2007      |
| Stipe length (c)    | Shorter lengths minimize resistance                        | @WatkinsJr.2010 |
| Frond margin (q)    | Thermoregulation                                           | @Nicotra2011    |
| Frond shape (q)     | Less divided fronds have lower rates of evapotranspiration | @Kluge2007      |
| Frond texture (q)   | Thicker leaves have lower rates of evapotranspiration      | @Givnish1987    |

**`r table("traits-used")`**. Fern traits used in this study. Letters in parentheses indicate trait type: b, binary (trait present or absent); c, continuous; q, qualitative. Qualitative traits were coded as binary for analysis (see Materials and Methods). For complete list of trait states for qualitative traits, see `r s_table("phy-sig-binary")`.

\newpage

```{r plot-prep}
# (color palettes are set in ms_functions.R)

# Define nicely formatted terms for printing variable names
resp_var_lookup <- tribble(
  ~resp_var, ~resp_var_pretty,
  "fd_obs_z", "SES of FD",
  "pd_obs_z", "SES of PD",
  "pe_obs_p_upper", "PE *p*-value",
  "rfd_obs_z", "SES of RFD",
  "richness", "Richness",
  "rpd_obs_z", "SES of RPD",
)

indep_var_lookup <- tribble(
  ~term, ~term_pretty,
  "temp", "Temperature",
  "precip", "Precipitation",
  "precip_season", "Precipitation seasonality",
  "I(temp^2)", "Temperature^2^",
  "percent_apo", "% apomictic taxa"
)

# Crop high-res map of Japan to spatial div results area
japan <- sf::st_crop(japan_shp, sf::st_bbox(biodiv_ferns_spatial)) %>%
  st_transform(4612)

# Format biodiv_ferns_spatial data for plotting
biodiv_ferns_spatial <-
  biodiv_ferns_spatial %>%
  # Set order of factor levels for plotting
  mutate(
    endem_type = factor(
      endem_type,
      levels = c("neo", "paleo", "not significant", "mixed", "super")),
    # can ignore warning about unknown levels, since they don't exist in all
    # variables but that's OK
    across(
      contains("signif"),
      ~fct_relevel(.x,
      "< 0.01",  "< 0.025", "not significant", "> 0.975", "> 0.99")
    )
  ) %>%
  st_transform(4612)
```

```{r japan-map-make}
# Plot map of Japan with labels of relevant places

# Set vectors of different areas to label
main_islands <- c("Hokkaido", "Shikoku", "Kyushu", "Honshu")
small_islands <- c("Okinawa", "Amami", "Ogasawara Islands", "Yakushima",
  "Izu Islands", "Iriomote", "Miyako")
other <- c("Tokyo")

# Modify dataframe to include formatting for printing label on map
japan_points <-
  japan_map_points %>%
  filter(is.na(lat_end)) %>%
  select(-lat_end, -lon_end) %>%
  mutate(
    label = case_when(
      # Print main islands in bold
      query %in% main_islands ~ glue::glue("**{query}**"),
      TRUE ~ query
    ),
    nudge = case_when(
      query == "Honshu" ~ "-0.5,-.7",
      query == "Tokyo" ~ "-1.2,0",
      query == "Shikoku" ~ "1.3,-.6",
      query == "Kyushu" ~ "1.6,-.4",
      query == "Miyako" ~ "1.5, 0",
      query == "Iriomote" ~ "1.8, 0",
      query == "Izu Islands" ~ "2.0, 0",
      query == "Yakushima" ~ "2.2, 0",
      query == "Ogasawara Islands" ~ "3.0, 0",
      query == "Izu Peninsula" ~ "2.0, -0.3",
      query == "Kii Peninsula" ~ "2.0, -0.5",
      query == "Ryukyu Islands" ~ "8.0, 1.0",
      query %in% small_islands ~ "2.0, 0",
      TRUE ~ "0,0"
    )
  ) %>%
  separate(nudge, c("nudge_lon", "nudge_lat"), sep = ",") %>%
  mutate(across(c(nudge_lon, nudge_lat), as.numeric)) %>%
  mutate(
    lon_adj = lon + nudge_lon,
    lat_adj = lat + nudge_lat
  )

japan_lines <- japan_map_points %>%
  filter(!is.na(lat_end))

# Make map
japan_map_plot <- ggplot(japan) +
  geom_sf(size = 0.25) +
  geom_segment(
    data = japan_lines,
    aes(x = lon, y = lat, xend = lon_end, yend = lat_end, linetype = query)
  ) +
  scale_linetype_manual(
    values = c("Tsugaru Strait" = "solid", "Tokara Strait" = "longdash")
  ) +
  geom_richtext(
    data = japan_points,
    aes(x = lon_adj, y = lat_adj, label = label),
    size = 2.5,
    fill = NA, label.color = NA, # remove background and outline
    label.padding = grid::unit(rep(0, 4), "pt") # remove padding)
  ) +
  geom_point(
    data = filter(japan_points, query == "Tokyo"),
    aes(x = lon, y = lat)
  ) +
  annotation_scale(location = "br", width_hint = 0.2) +
  map_theme() +
  theme(legend.position = "none")
```

```{r japan-map-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("japan-map"), "pdf"),
  japan_map_plot, width = 5, height = 5, units = "in")
```

```{r japan-map-print, fig.width = 5, fig.height = 5, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
japan_map_plot
```

**`r figure("japan-map") %>% str_to_upper()`**. Map of Japan showing names of places mentioned in the text. Names of the four main islands in bold. The Ryukyu Islands include Yakushima, Amami, Okinawa, Miyako, and Iriomote. Solid line shows location of Tsugaru Strait; dashed line shows Tokara Strait.

`r pagebreak_pdf()`

```{r raw-diversity-make}
# Richness
a <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = richness), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**A** Richness</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "No. taxa") +
  scale_fill_scico(palette = "lajolla") +
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  map_theme()

# Raw PD
b <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = pd_obs), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**B** PD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "Relative\nbranch\nlength") +
  scale_fill_scico(palette = "lajolla") +
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  map_theme() +
  theme(axis.text.y = element_blank())

# Raw FD
c <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = fd_obs), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**C** FD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "Relative\nbranch\nlength") +
  scale_fill_scico(palette = "lajolla") +
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  annotation_scale(location = "tl", width_hint = 0.2, pad_y = unit(0.8, "cm")) +
  map_theme() +
  theme(axis.text.y = element_blank())

raw_diversity_plot <- a + b + c +
  plot_layout(nrow = 1)
```

```{r raw-diversity-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("raw-diversity"), "pdf"),
  width = 7.25, height = 3.4, units = "in", raw_diversity_plot)
```

```{r raw-diversity-print, fig.width = 7.25, fig.height = 3.4, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
raw_diversity_plot
```

**`r figure("raw-diversity") %>% str_to_upper()`**. Raw (observed) biodiversity metrics of the ferns of Japan. (A) Raw taxonomic diversity. (B) Raw phylogenetic diversity. (C) Raw functional diversity. For (B) and (C), raw branchlengths were transformed to relative values by dividing each branchlength by the sum of all branchlengths.

`r pagebreak_pdf()`

```{r rand-diversity-make}
# PD
a <- ggplot() +
  # white base map, no outline
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  # colored cells
  geom_sf(data = biodiv_ferns_spatial, aes(fill = pd_signif), color = "transparent") + #nolint
  # light grey coastline
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**A** PD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  scale_fill_manual(values = signif_cols) +
  map_theme() +
  theme(
    axis.text.x = element_blank(),
    legend.title = element_blank()
  )

# RPD
b <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = rpd_signif), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**B** RPD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  map_theme() +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    legend.title = element_blank()
  )

# FD
c <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = fd_signif), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**C** FD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
  map_theme() +
  theme(legend.title = element_blank())

# RFD
d <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = rfd_signif), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**D** RFD</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
  annotation_scale(location = "br", width_hint = 0.2) +
  map_theme() +
  theme(
    axis.text.y = element_blank(),
    legend.title = element_blank()
  )

rand_diversity_plot <- a + b + c + d +
  plot_layout(ncol = 2, nrow = 2, guides = "collect") &
  theme(legend.position = "bottom")
```

```{r rand-diversity-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("rand-diversity"), "pdf"),
  rand_diversity_plot, width = 7.25, height = 8, units = "in")
```

```{r rand-diversity-print, fig.width = 7.25, fig.height = 9, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
rand_diversity_plot
```

**`r figure("rand-diversity") %>% str_to_upper()`**. Results of randomization tests for phylogenetic and functional diversity of the ferns of Japan. (A) Phylogenetic diversity (PD). (B) Relative phylogenetic diversity (RPD). (C) Functional diversity (FD). (D) Relative functional diversity (RFD). For each metric, raw values were compared to a null distribution of 999 random values, and the *P*-value calculated using a two-tailed test. For details, see Materials and Methods.

`r pagebreak_pdf()`

```{r coeff-make}
### Make model coefficients plot ###

# Format labels for printing
resp_var_lookup <-
  tibble(
    resp_var = c("richness", "fd_obs_z", "pd_obs_z", "rfd_obs_z", "rpd_obs_z"),
    resp_var_print = c(
      "Richness", "SES of FD", "SES of PD", "SES of RFD", "SES of RPD")
  )

indep_var_lookup <-
  tibble(
    term = c("lat_area", "I(temp^2)", "null_model", "percent_apo", "precip",
      "precip_season", "temp"),
    term_print = c("Area", "Temp^2", "Null model", "% apomictic", "Precip.",
      "Precip.\nseasonality", "Temp.")
  ) %>%
  # Include "plain" version of term for arranging alphabetically
  mutate(term_plain = str_replace_all(term, fixed("I(temp^2)"), "temp^2"))

model_type_lookup <-
  tibble(
    model_type = c("temp", "percent_apo"),
    model_type_print = c("env. model", "repro. model")
  )

# Specify subplot order
model_subplot_order <- tribble(
  ~resp_var, ~indep_var, ~order,
  "richness", "temp", 1,
  "pd_obs_z", "temp", 2,
  "rpd_obs_z", "temp", 3,
  "fd_obs_z", "temp", 4,
  "rfd_obs_z", "temp", 5,
  "pd_obs_z", "percent_apo", 6,
  "rpd_obs_z", "percent_apo", 7
)

# Format likelihood ratio test results
lrt_signif <-
  lrt_results %>%
  # add model type (temp = environmental model, or percent_apo = repro model)
  tidyr::extract(full_formula, "model_type", regex = " ~ ([^ ]+) ") %>%
  select(-null_formula) %>%
  select(resp_var, model_type, term = comparison, p_value) %>%
  # Add column for significance
  mutate(signif = p_value < 0.05)

# Format parameter values for plotting
model_params_for_plot <-
  model_params %>%
  # add model type
  tidyr::extract(formula, "model_type", regex = " ~ ([^ ]+) ") %>%
  # drop intercept
  filter(term != "(Intercept)") %>%
  # add LRT results and names formatted for printing
  left_join(lrt_signif, by = c("resp_var", "term", "model_type")) %>%
  left_join(resp_var_lookup, by = "resp_var") %>%
  left_join(indep_var_lookup, by = "term") %>%
  left_join(model_type_lookup, by = "model_type") %>%
  # Make coefficient with greatest abs t-value bold
  group_by(resp_var, model_type) %>%
  mutate(
    term_print = case_when(
      abs(t_value) == max(abs(t_value)) ~ glue::glue("**{term_print}**"),
      TRUE ~ term_print
    ),
    term_print = factor(term_print)
  ) %>%
  ungroup() %>%
  assert(not_na, matches("print|plain"))

# Define helper function to make a subplot showing coefficients
plot_coeff <- function (df) {

  plot_data <- df %>%
    mutate(term_print = fct_reorder(term_print, rank(term_plain)) %>% fct_rev)

  ggplot(plot_data, aes(x = t_value, y = term_print, fill = signif)) +
    geom_vline(xintercept = 0, linetype = "dotted") +
    geom_col() +
    # scale_x_continuous(n.breaks = 4) +
    scale_x_continuous() +
    scale_fill_manual(
      values = c("TRUE" = "grey20", "FALSE" = "grey70"),
      labels = c("TRUE" = "*p* < 0.05", "FALSE" = "*p* > 0.05")) +
    labs(
      subtitle = glue("{unique(plot_data$resp_var_print)},\n{unique(plot_data$model_type_print)}"), #nolint
      fill = "Significance") +
    theme_bw(base_size = 9) +
    theme(
      axis.title = element_blank(),
      axis.text.y = element_markdown(),
      legend.text = element_markdown(),
      panel.grid.minor = element_blank(),
      legend.position = "none")
}

# Loop across data and make a plot of each
coeff_subplots <-
  model_params_for_plot %>%
  group_by(resp_var, model_type) %>%
  nest() %>%
  ungroup %>%
  left_join(
    model_subplot_order, by = c("resp_var", model_type = "indep_var")) %>%
  arrange(order) %>%
  mutate(plot = map(data, plot_coeff)) %>%
  pull(plot)

# We need a legend from one of these
coeff_subplots[[7]] <- coeff_subplots[[7]] + theme(legend.position = "right")

# Generate final plot
coeff_plot <- wrap_plots(coeff_subplots) +
  guide_area() +
  plot_layout(guides = "collect") +
  plot_annotation(
    tag_levels = "A",
    tag_prefix = "",
    tag_suffix = "") &
  theme(plot.tag = element_text(face = "bold"))
```

```{r coeff-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("coeff"), "pdf"),
  coeff_plot, width = 7.25, height = 5, units = "in")
```

```{r coeff-print, fig.width = 7.25, fig.height = 5, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
coeff_plot
```

**`r figure("coeff") %>% str_to_upper()`**. Regression coefficient *t*-values (coefficient divided by standard error) of general mixed models for the effect of environment and reproductive mode on biodiversity metrics in the ferns of Japan. (A)--(E) Environmental models (models including temperature). (F, G) Reproductive models (models including % apomictic taxa). Statistical significance assessed with a likelihood ratio test (LRT) between the full model and a model with the focal independent variable removed (null model); *P*-value indicates probability of no difference between the full and null model. Name of response variable and model type indicated above each subplot. Independent variable with greatest absolute *t*-value in bold for each model.

```{r modelfit-make}
# Make plot showing model fits

# Extract dataset used for model, including significance columns
biodiv_ferns_used_for_model <- biodiv_ferns_spatial %>%
  # drop geometry (only need centroids for modeling)
  st_set_geometry(NULL) %>%
  as_tibble() %>%
  # filter to only sites used in model
  inner_join(
    biodiv_ferns_cent %>% as_tibble() %>% select(grids),
    by = "grids"
  )

# Generate final plot
modelfit_plot_list <-
  model_fits %>%
  mutate(
    signif_var = str_replace_all(resp_var, "_obs_z", "_signif"),
    resp_var_print = case_when(
      resp_var == "richness" ~ "Richness (no. spp.)",
      TRUE ~ str_to_upper(resp_var) %>% word(1, 1, "_") %>% paste("SES of", .)
    ),
    model_data = list(biodiv_ferns_used_for_model),
    signif_cols = list(signif_cols)) %>%
  left_join(model_subplot_order, by = c("resp_var", "indep_var")) %>%
  arrange(order) %>%
  pmap(plot_fits)

# Need one subplot with a legend
modelfit_plot_list[[2]] <- modelfit_plot_list[[2]] +
  theme(legend.position = "right") +
  labs(color = "Significance") +
  # Increase size of points in legend
  guides(colour = guide_legend(override.aes = list(size = 1.5)))

modelfit_plot <-
  wrap_plots(modelfit_plot_list) +
  guide_area() +
  plot_layout(guides = "collect") +
  plot_annotation(
    tag_levels = "A",
    tag_prefix = "",
    tag_suffix = "")  &
  theme(plot.tag = element_text(face = "bold"))
```

```{r modelfit-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("modelfit"), "pdf"),
  modelfit_plot, width = 7.25, height = 5, units = "in")
```

```{r modelfit-print, fig.width = 7.25, fig.height = 5, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
modelfit_plot
```

**`r figure("modelfit") %>% str_to_upper()`**. Relationship between biodiversity metrics and selected predictor variables in the ferns of Japan. (A)--(E) Environmental models (models including temperature). (F, G) Reproductive models (models including % apomictic taxa). Ribbon shows 95% confidence interval of model. Line fit with the focal predictor variable while averaging over other predictors. For phylogenetic diversity (PD), relative PD (RPD), functional diversity (FD), and relative FD (RFD), the standard effect size (SES) was calculated by comparing raw values to a null distribution of 999 random values, and significance (*P*-value) calculated using a two-tailed test (see Materials and Methods).

`r pagebreak_pdf()`

```{r endemism-make}
endemism_plot <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial, aes(fill = endem_type), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  scale_fill_manual(values = canape_cols) +
  annotation_scale(location = "br", width_hint = 0.2) +
  map_theme() +
  theme(legend.title = element_blank())
```

```{r endemism-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("endemism"), "pdf"),
  endemism_plot, width = 5, height = 5, units = "in")
```

```{r endemism-print, fig.width = 5, fig.height = 5, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
endemism_plot
```

**`r figure("endemism") %>% str_to_upper()`**. Phylogenetic endemism of the ferns of Japan measured using CANAPE (categorical analysis of neo- and paleo-endemism). 'neo' indicates areas with an overabundance of rare short branches; 'paleo' indicates areas with an overabundance of rare long branches; 'mixed' indicates significantly endemic sites that have neither an overabundance of short nor long branches; 'super' indicates highly significantly endemic sites that have neither an overabundance of short nor long branches. For details, see Materials and Methods.

`r pagebreak_pdf()`

```{r bioregions-make}
# Taxonomic bioregions
a <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial_lumped, aes(fill = taxonomic_cluster), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**A** Taxonomic</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "Bioregion") +
  scale_fill_manual(values = bioregion_cols) +
  map_theme()

# Phylogenetic bioregions
b <- ggplot() +
  geom_sf(data = japan, fill = "white", color = "transparent", lwd = coast_lwd) + #nolint
  geom_sf(data = biodiv_ferns_spatial_lumped, aes(fill = phylo_cluster), color = "transparent") + #nolint
  geom_sf(data = japan, fill = "transparent", color = coast_col, lwd = coast_lwd) + #nolint
  annotate(
    geom = "richtext", y = Inf, x = -Inf,
    label = "<span style='color:black'>**B** Phylogenetic</span>",
    hjust = "inward", vjust = "inward", color = "white") +
  labs(fill = "Bioregion") +
  scale_fill_manual(values = bioregion_cols) +
  annotation_scale(location = "br", width_hint = 0.2) +
  map_theme() +
  theme(axis.text.y = element_blank())

bioregions_plot <- a + b +
  plot_layout(ncol = 2)
```

```{r bioregions-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("bioregions"), "pdf"),
  bioregions_plot, width = 7.25, height = 4.3, units = "in")
```

```{r bioregions-print, fig.width = 7.25, fig.height = 4.3, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
bioregions_plot
```

**`r figure("bioregions") %>% str_to_upper()`**. Bioregions of the ferns of Japan. (A) Taxonomic bioregions. (B) Phylogenetic bioregions. Bioregions determined by clustering taxonomic (Sørensen) or phylogenetic (PhyloSor) distances between grid cells. Bioregions not consisting of more than `r english2(bioregion_cutoff)` grid cells each are lumped into the "Other" category.

`r pagebreak_pdf()`

```{r div-by-tax-region-make}
data <-
  biodiv_ferns_spatial %>%
  filter(taxonomic_cluster %in% c(1, 2, 3, 4)) %>%
  as_tibble %>%
  select(
    pd_obs_z, rpd_obs_z, fd_obs_z, rfd_obs_z,
    pe_obs_p_upper, taxonomic_cluster)

a <- ggplot(data, aes(x = taxonomic_cluster, y = pd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = bioregion_cols, breaks = 1:4) +
  geom_boxplot(
    color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES PD",
    x = "Cluster",
    color = "Bioregion"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

b <- ggplot(data, aes(x = taxonomic_cluster, y = rpd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = bioregion_cols, breaks = 1:4) +
  geom_boxplot(
    color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RPD",
    x = "Cluster",
    color = "Bioregion"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

c <- ggplot(data, aes(x = taxonomic_cluster, y = fd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = bioregion_cols, breaks = 1:4) +
  geom_boxplot(
    color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES FD",
    x = "Cluster",
    color = "Bioregion"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

d <- ggplot(data, aes(x = taxonomic_cluster, y = rfd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = bioregion_cols, breaks = 1:4) +
  geom_boxplot(
    color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RFD",
    x = "Cluster",
    color = "Bioregion"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.title.x = element_blank()
  )

e <- ggplot(data, aes(x = taxonomic_cluster, y = pe_obs_p_upper)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = bioregion_cols, breaks = 1:4) +
  geom_boxplot(
    color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "PE *p*-score",
    x = "Cluster",
    color = "Bioregion"
  ) +
  theme_bw() +
  theme(
    axis.title.y = element_markdown(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
  )

div_by_tax_region_plot <- a + b + c + d + e +
  plot_layout(ncol = 2) +
  plot_annotation(
    tag_levels = "A",
    tag_prefix = "",
    tag_suffix = ""
  ) &
  theme(
    legend.position = "none",
    plot.tag = element_text(face = "bold"))
```

```{r div-by-tax-region-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("div-by-tax-region"), "pdf"),
  div_by_tax_region_plot, width = 7.25, height = 7, units = "in")
```

```{r div-by-tax-region-print, fig.width = 7.25, fig.height = 7, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
div_by_tax_region_plot
```

**`r figure("div-by-tax-region") %>% str_to_upper()`**. Phylogenetic and morphological diversity of the ferns of Japan by taxonomic bioregion. (A) Standard effect size (SES) of phylogenetic diversity (PD). (B) SES of relative PD (RPD). (C) SES of functional diversity (FD). (D) SES of relative FD (RFD). (E) *P*-score of observed phylogenetic endemism (PE) relative to 999 random communities.

`r pagebreak_pdf()`

```{r conserv-status-make}
# Add total protected area (medium + high)
signif_cells_protected_area_total <- #nolint
  signif_cells_protected_area %>%
  group_by(metric) %>%
  summarize(percent_protected = sum(percent_protected), .groups = "drop") %>%
  mutate(status = "total")

total_percent_protected_for_plot <- #nolint
  total_percent_protected %>%
  select(status, percent_protected) %>%
  mutate(status = str_to_sentence(status))

# Plot protected areas
a <- signif_cells_protected_area %>%
  select(metric, status, percent_protected) %>%
  bind_rows(signif_cells_protected_area_total) %>%
  mutate(
    metric = case_when(
      metric == "pd" ~ "PD",
      metric == "fd" ~ "FD",
      metric == "pe" ~ "PE",
      metric == "richness" ~ "Richness",
    ),
    status = str_to_sentence(status)) %>%
  mutate(metric = factor(metric, levels = c("PE", "FD", "PD", "Richness"))) %>%
  ggplot(aes(x = percent_protected, y = metric, fill = status)) +
  geom_col(position = position_dodge()) +
  geom_vline(
    data = total_percent_protected_for_plot,
    aes(xintercept = percent_protected, linetype = status),
    color = "grey30"
  ) +
  scale_linetype_manual(
    values = protection_lines,
    guide = guide_legend(
      direction = "horizontal",
      title.position = "top",
      label.position = "left"
    )) +
  scale_fill_manual(
    values = protection_cols,
    guide = guide_legend(
      direction = "horizontal",
      title.position = "top",
      label.position = "left"
    )) +
  scale_x_continuous(
    labels = function(x) scales::percent(x, accuracy = 1),
    expand = expansion(mult = c(0, .05))) +
  labs(
    x = "Percent protected",
    y = "Biodiversity metric",
    fill = "Protection status",
    linetype = "Protection status"
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "bottom"
  )

# Plot areas in danger from herbivory by deer
b <- signif_cells_deer_danger %>%
  mutate(
    metric = case_when(
      metric == "pd" ~ "PD",
      metric == "fd" ~ "FD",
      metric == "pe" ~ "PE",
      metric == "richness" ~ "Richness",
    ),
    range = str_to_sentence(range)) %>%
  mutate(metric = factor(metric, levels = c("PE", "FD", "PD", "Richness"))) %>%
  ggplot(aes(x = percent_danger, y = metric, fill = range)) +
  geom_col(position = position_dodge()) +
  geom_vline(
    data = total_percent_deer_danger %>% mutate(range = str_to_sentence(range)),
    aes(linetype = range, xintercept = percent_danger),
    color = "grey30"
  ) +
  scale_linetype_manual(
    values = deer_lines,
    guide = guide_legend(
      direction = "horizontal",
      title.position = "top",
      label.position = "left"
    )) +
  scale_fill_manual(
    values = deer_cols,
    guide = guide_legend(
      direction = "horizontal",
      title.position = "top",
      label.position = "left"
    )) +
  scale_x_continuous(
    labels = function(x) scales::percent(x, accuracy = 1),
    expand = expansion(mult = c(0, .05))) +
  labs(
    x = "Percent area with deer",
    y = "Biodiversity metric",
    fill = "Range data source",
    linetype = "Range data source"
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.title.y = element_blank(),
    legend.position = "bottom"
  )

conserv_status_plot <- a + b +
  plot_layout(ncol = 2) +
  plot_annotation(
    tag_levels = "A",
    tag_prefix = "",
    tag_suffix = ""
  ) &
  theme(plot.tag = element_text(face = "bold"))
```

```{r conserv-status-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
  result_file(figure("conserv-status"), "pdf"),
  conserv_status_plot, width = 7.25, height = 4, units = "in")
```

```{r conserv-status-print, fig.width = 7.25, fig.height = 4, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
conserv_status_plot
```

**`r figure("conserv-status") %>% str_to_upper()`**. Conservation status and threat due to herbivory by deer in areas with high levels of biodiversity for ferns of Japan. (A) Percent of land area with medium or high protected status for grid cells with significantly high biodiversity by protection status. (B) Percent of land area occupied by deer for grid cells with significantly high biodiversity by deer range data source (1978 survey, 2003 survey, or range estimated from model based on 2003 survey data). Biodiversity metrics include taxon richness, phylogenetic diversity (PD), functional diversity (FD), and phylogenetic endemism (PE). Significance of PD, FD, and PE assessed by a one-tailed test comparing observed values to a null distribution of 999 random values; for richness, grid cells in the top 5% considered highly diverse (see Materials and Methods). For (A), vertical lines indicate total percent area protected across Japan (baseline protection rate). For (B), vertical lines indicate total percentage of land area occupied by deer across Japan (baseline threat rate).
