---
title: "Exploring dimensions of biodiversity in Japanese ferns"
bibliography: "main_library.bib"
mainfont: Roboto-Regular.ttf
fontsize: 11pt
mainfontoptions: 
- BoldFont=Roboto-Bold.ttf
- ItalicFont=Roboto-Italic.ttf
- BoldItalicFont=Roboto-BoldItalic.ttf
output: 
  bookdown::pdf_document2:
    citation_package: none # Use pandoc default, otherwise csl can't be used for reference formatting
    toc: no
    number_sections: no
    fig_caption: yes
    keep_tex: no
    latex_engine: xelatex
    pandoc_args: [ "--csl", "journal-of-biogeography.csl"]
    includes:
      in_header: newphyt.sty
header-includes: 
  \usepackage{float}
  \makeatletter\renewcommand*{\fps@figure}{H}\makeatother
editor_options: 
  chunk_output_type: console
# nocite: | 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, results = "hide", cache = FALSE)
```

```{r load-data, cache = FALSE}
# NOTE: This Rmd is meant to be knit as part of the drake plan (R/plan.R).
# If using interactively, run `source("_drake.R")` first.

loadd(list = c(
  "occ_point_data_ferns",
  "occ_point_data_ferns_unfiltered",
  "biodiv_ferns_spatial",
  "k_taxonomy",
  "k_phylogeny",
  "signif_cells_protected_area",
  "protected_areas",
  "japan_shp"
),
cache = ja_fern_cache)

phylo_cache <- new_cache(here::here("phylo_cache"))

loadd(list = c(
  "plastome_calibration_dates",
  "plastome_tree"
),
cache = phylo_cache)

# Load captions
source(here::here("ms/captions.R"))
```

<!-- Remove page numbers -->
\pagenumbering{gobble}

<!-- Set left justification -->
\raggedright

<!-- Don't indent -->
\setlength{\parindent}{0ex}

Joel H. Nitta^1^, Brent D. Mishler^2^, Wataru Iwasaki^1^, and Atsushi Ebihara^3^

^1^Department of Biological Sciences, Graduate School of Science, The University of Tokyo, 2-11-16 Yayoi, Bunkyo-ku, Tokyo 113-0032, Japan

^2^University and Jepson Herbaria, University of California, Berkeley, CA 94720‐2465, USA

^3^Department of Botany, National Museum of Nature and Science, 4-1-1 Amakubo, Tsukuba-shi, Ibaraki, 305-0005, Japan  

Author for correspondence:  
*Joel H. Nitta*  
*Email: joelnitta@gmail.com*

\clearpage

## Abstract

The ferns of Japan are one of the best-characterized regional floras in the world due to decades of intensive collection and taxonomic revision. We combined phylogenetic, functional, and environmental datasets to explore the spatial distribution of biodiversity in the ferns of Japan. We observe three major patterns in the spatial distribution of biodiversity: that of taxonomic richness, which is highest at mid-latitudes, that of phylogenetic and functional diversity, which is highest in small southern islands, and that of relative phylogenetic and functional diversity, which is high at both high and low latitudes, and low at mid-latitudes.

**Key words:** biodiversity, biogeography, phylogenetic diversity. 

\clearpage

## Introduction

Characterizing the spatial distribution of biodiversity is a major goal of evolutionary biology with two equally important aspects: to understand the processes generating it, and to conserve it. The vast majority of studies seeking to understand the spatial distribution of biodiversity focus on species richness. Hotspots, or areas of exceptionally high richness, have received particular attention, resulting in a widely-recognized set of 34 terrestrial hotspots and motivating conservation strategies to preserve them. However, richness alone cannot provide a complete picture of biodiversity. All organisms are related by descent from a common ancestor, and these evolutionary relationships must be taken into account to obtain a full understanding of biodiversity.

Presumably, the main reason phylogeny has not been taken into account more prominently in biodiversity studies is because data (DNA sequences) and analytical tools have only become available relatively recently. These tools and datasets now make it possible to analyze other dimensions of biodiversity, such as phylogenetic diversity and phylogenetic endemism. Better ways to characterize biological regions ("bioregions") are also available, rather than relying on ad-hoc characterizations. Furthermore, incorporating these two frameworks---the categorization of areas into phyloregions with analysis of over/under dispersion of biodiversity---can provide particularly powerful insights into the processes structuring biodiversity and suggest conservation priorities.

However, a comprehensive understanding of the relationships between richness and other metrics requires densely sampled data or can be lead astray.

Ferns of Japan are excellent model system because they have been the target of intense botanical interest for several decades and are very densely sampled. Japan is situated along a latitudinal gradient spanning highly season temperate areas in the north to subtropical, mostly aseasonal islands in the south.

Here we leverage this exceptionally rich dataset to analyze the geographic distribution of biodiversity in detail. We ask the following questions in our study system, the ferns of Japan: 1. How is biodiversity distributed? 2. How is biodiversity structured? 3. How well is biodiversity protected?

## Materials and Methods

### Study Site

Our data include all ferns known to occur within the political boundaries of Japan (20° to 45° N, 122° to 153° E). Japan consists of four main islands and thousands of smaller ones. The climate is subtropical in the south and highly seasonal in other parts of the country.

### Occurrence data

We used a list of all non-hybrid fern specimens housed at herbaria in Japan. We chose to use this over other sources (e.g., GBIF), because of its high taxonomic quality: all identifications are standardized to a common taxonomy, the Japan Green List v. 1.01 (http://www.rdplants.org/gl/). Furthermore, the overall sampling completeness of this dataset is extremely high (`r s_figure("sampling-curve")`). The original list comprises `r nrow(occ_point_data_ferns_unfiltered) %>% scales::number(big.mark = ",")` specimens representing `r occ_point_data_ferns_unfiltered %>% pull(taxon) %>% n_distinct()` taxa (species, subspecies, and varieties). We filtered out any occurrences not within a map of standard grid squares defined by the Statistics Bureau of Japan (http://www.stat.go.jp/english/data/mesh/05.html). The final cleaned occurrence dataset included `r nrow(occ_point_data_ferns) %>% scales::number(big.mark = ",")` specimens (`r occ_point_data_ferns %>% pull(taxon) %>% n_distinct()` taxa). Hereafter, "taxa" refers to species, subspecies, and varieties unless otherwise indicated.

A necessary step in any analysis of biodiversity is to set the grain size used to define communities. This decision must consider that smaller grain size is needed to detect environmental effects, while larger grain size is needed to ensure adequate species sampling. We tested using 0.1, 0.2, 0.3, and 0.4 degree grid-cells for defining communities (co-occurring species). At each size, species occurrences were converted into a presence-absence community matrix (a species was considered present if at least one specimen was recorded in that grid cell). We calculated sampling redundance (1 - richness/number of specimens) [@Garcillan2003] to quantify adequacy of sampling. Preliminary analysis indicated that 0.2 degree grid-cells are optimal for our dataset (`r s_figure("grain-size")`). We filtered out any cells with redundancy < 0.1 (mostly cells along the coast with very low land area) and used this for subsequent analyses.

### Trait data

We used traits originally compiled for identification of ferns and lycophytes of Japan using LUCID software [@Ebihara2016b; @Ebihara2017; @Ebihara2019b]. Quantitative traits were measured on 10 randomly selected specimens per species, and qualitative traits determined by observing voucher specimens [@Ebihara2016b; @Ebihara2017]. All qualitative traits were measured in binary format; \eg, a single trait with three states was formatted as three binary traits. Quantitative traits were measured in fertile (i.e., spore-bearing) and sterile fronds separately originally, then averaged. From the original trait list, we selected only traits with putatative ecological function, and removed any that had Pearson correlation coefficient >0.6 or fewer than three observations of a given trait state (`r table("traits-used")`).

### Phylogenetic analysis

Sequences for the plastid, coding *rbcL* gene are available for ca. 97% of the Japanese fern flora [@Ebihara2019b]. However, using this dataset alone for phylogeny suffers two drawbacks: it is not densely sampled enough to resolve many internal nodes, and it lacks many lineages that have fossils availabsle for molecular dating. Because one goal of the study is to understand the distribution of paleo- vs. neo-endemic areas (defined in units of time), we require an ultrametric tree. To obtain a maximally sampled, robust phylogenetic tree, we combined the *rbcL* data of @Ebihara2019b with a globally sampled plastid dataset including all fern sequences on GenBank for four plastid genes (*rbcL*, *atpB*, *aptA*, and *rps4*) and 66 other coding, single-copy plastid genes extracted from 100 complete fern plastomes (Nitta et al., in prep) (`r ape::Ntip(plastome_tree) %>% scales::number(big.mark = ",")` OTUs total). We conducted phylogenetic analysis with IQTREE [@Nguyen2015] and molecular dating with treePL [@Smith2012] using `r nrow(plastome_calibration_dates)` fossil calibration points after @Testo2016a. We then trimmed the tree to Japanese taxa only and used this for all further analyses.

### Phylogenetic signal

We tested for phylogenetic signal in continuous traits using two alternative metrics, Blomberg's *K* [@Blomberg2003] and Pagel's lambda [@Pagel1999]. *K* describes the ratio between the observed variance in a trait vs. the amount of variance expected under Brownian Motion (BM); lambda is a scaling parameter used to transform the phylogeny such that trait values fit those most closely expected under BM. Both lambda and *K* = 1 under BM; lower values indicate less signal (overdispersion) and higher values indicate more signal (clustering). We measured *K* and lambda using the \function{phylosig} function in the R package \package{phytools} [@Revell2012], which also calculates significance based on a randomization test for *K* and a likelihood ratio test for lambda. We tested for phylogenetic signal in qualitative (coded as binary) traits using Fritz and Purvis' *D* [@Fritz2010] with the \function{phylo.d} function in the R package \package{caper} [@Orme2018], which also assesses significance by comparing the observed value to a null distribution of random values and a simulated distribution of values expected under BM. 

### Analysis of biodiversity

We measured taxonomic diversity using taxonomic richness.

We measured phylogenetic diversity (PD) using the Faith's PD (the total branch length connecting all terminal taxa in a community, including the root of the phylogeny) [@Faith1992] with the \function{PD} function in the R package \package{phyloregion}. To detect areas with unusually long or short branch lengths, we measured relative phylogenetic diversity (RPD), which is a ratio of observed PD to PD measured on a tree where all branch lengths have been set to equal length [@Mishler2014].

We measured functional diversity (FD) following the method of @Petchey2002, which is a functional analog of PD. We first calculated distances using Gower's method [@Gower1971] on the trait data with the \function{gowdis} function in the R package \package{FD} [@Laliberte2011]. We weighted quantitative and qualitative traits equally, weighted each binary component trait of each qualitative trait equally, and log-transformed and scaled quantitative traits prior to calculating distances. We then used the distances to build a trait dendrogram using the \function{hclust} function in the \package{stats} R package [@RCoreTeam2020], and used the dendogram in place of a phylogenetic tree to calculate PD and RPD as described above.

We measured phylogenetic endemism (PE) [@Rosauer2009], which is the sum of all branches at a site inversely weighted by their range, using the \function{phylo{\_}endemism} function in the R package \package{phyloregion}. 

Observed values of PD, FD, and PE are known to be closely related to species richness: as more taxa are drawn at random without replacement from a given set, it becomes increasingly unlikely to draw a taxon that is distantly related (or functionally dissimilar) to the others. Therefore, we determined whether the observed metrics are more or less diverse than random for a given number of taxa by conducting a randomization test. We generated 999 random communities using the "independent swap" algorithm of @Gotelli2000, which conserves richness per site and species occurrence frequencies while randomizing species identities, then compared the observed value to the distribution of random values. Statistical significance was assessed using a two-tailed test with alpha = 0.5: observed values > 97.5% or < 0.25% were considered significant.

We tested whether endemic areas have an over-representation of old (paleoendemic) or new (neoendemic) lineages using categorical analysis of neo- and paleo-endemism (CANAPE), which involves measuring PE with an alternate tree where all branch lengths are equal, then comparing the ratio between raw PE and alternative PE (RPE) [@Mishler2014]. Low RPE indicates short branchlengths (neoendemism) and high RPE indicates long branchlengths (paleoendemism). Since all of these measures are affected by species richness, categorization of endemism is done using the rank of each observed variable to those calculated for the null distribution. We used the same set of null communities as for SES of PE, and carried out CANAPE using custom R scripts.

It is possible that endemism may be affected by species that are actually wide-ranging, but only have a small range in Japan (the "border effect"). This is especially expected for southern, sub-tropical islands, which harbor many tropical species with small ranges in Japan but wider ranges outside of the country. To account for this, we also calculated all endemism-related metrics using a dataset including only species endemic to Japan.

### Analysis of bioregions

We analyzed bioregions using the framework of @Daru2020 and their R package \package{phyloregion}. Briefly, this involves first calculating beta diversity (change in composition between communities) using either species names (taxonomic bioregions) or phylogenetic distances (phylogenetic bioregions). For taxonomic bioregions, we calculated the turnover in species between sites due to species replacement [@Baselga2012] with the Sørensen index. For phylogenetic bioregions, we calculated phylogenetic dissimilarities based on the PhyloSor index [@Bryant2008]. The distances are used to construct a dendrogram, which is then split into into *k* bioregions. As the optimal number of bioregions cannot be known *a-priori*, we tested values of *k* from 1 to 20.

We compared patterns in biodiversity between bioregions by calculating the standard effect size (SES) of biodiversity metrics. SES measures how extreme the observed value (*obs*) is relative to the null distribution (*null*): 

$SES = \frac{obs - mean(null)}{sd(null)}$

We compared SES of PD, RPD, FD, and RFD across bioregions using FIXME: add this when figure out spatial-corrected statistical method to use.

### Assessment of conservation status

We downloaded SHP files of protected areas in Japan from the Japan Ministry of the Environment (https://www.biodic.go.jp/biodiversity/activity/policy/map/map17/index.html). We only considered areas that include protection of plants, and categorized protected areas as either "high" (no human activities allowed at all) or "normal" (some human activities allowed by permit). We cropped grid cells to only include land area, then calculated the percent of protected area of each status type within grid cells with high levels of biodiversity as measured with PD, FD, PE, or taxonomic richness. For PD, FD, and PE, significance was assessed with a one-tailed test at alpha = 0.05 against the 999 null communities described above. For taxonomic richness, cells ranked in the top >5% were considered highly diverse.

### Data accessibility

Code to replicate all analyses, figures, and this manuscript are available at https://github.com/joelnitta/japan_ferns_biogeo. Raw and processed data are openly available from the Dryad repository at http://doi.org/ADD_DOI_HERE_WHEN_ASSIGNED.

## Results

### Datasets

The topology of the phylogenetic tree was in general agreement with other recent, large plastid fern phylogenies. Ferns were recovered as monophyletic, as were all families and genera; taxonomy follows PPGI for the family level and the Japan Green List for the genus level and below. The age of the fern clade is estimated to be _ my.

The distribution of specimen collection locations is highly skewed, with >3,000 specimens per grid cell northwest of Tokyo in Saitama prefecture, and many fewer elsewhere (`r s_figure("abundance")`A). Despite this, sampling redundancy is generally high throughout the country, with areas of low redundancy in Hokkaido and some coastal areas and small islands (`r s_figure("abundance")`B). The grid cells that had to be excluded from the analysis due to low redundancy were mostly from Hokkaido.

### Phylogenetic signal

Results of phylogenetic signal analysis are given in Table _. All quantitative traits showed significant phylogenetic signal, but estimates of signal strength differed between K and lambda: all lambda values were > 0.95, indicating evolution by BM; however, values of K were < 0.06, indicating less phylogenetic signal than expected under BM. A minority of quantitative traits (coded as binary) showed as much or more phylogenetic signal as expected under BM (D =< 0); others had D > 0.

### Observed diversity

Taxonomic richness is lowest on the northernmost island of Hokkaido, then increases with decreasing latitude until reaching a maximum in southern Kyushu, then decreases again in smaller islands further south (`r figure("raw-diversity")`A). Observed values of PD and FD were highly correlated with richness as expected (`r figure("raw-diversity")`B, C), showing an initial steep slope that gradually starts to level off at higher richness (`r s_figure("raw-biplot")`A, C). Observed values of RPD and RFD were also closely related to richness, showing high variance at low species richness, and low variance at high species richness (`r s_figure("raw-biplot")`B, D).

### Randomization tests for phylogenetic and functional diversity

Grid cells with significantly low PD were almost all located on Honshu, the main island of Japan. A smaller number of cells had significantly high PD, and these were located mostly on smaller southern (Amami, Ryukyu, Izu, Ogasawara) islands or coastal areas (Izu Peninsula, Kii Peninsula). The area near Yakushima includes cells with both significantly high and low PD. Grid cells with significantly low RPD were mostly observed in southern Honshu, particularly in and around Kyushu. Grid cells with significantly high RPD are located both in the southern islands and northern Honshu and Hokkaido. Results of the randomization test for functional diversity were broadly similar to phylogenetic diversity, but with more significant cells for FD than PD and fewer significant cells for RFD than RPD.

### Categorical analysis of neo- and paleo-endemism

Nearly all grid cells in islands south of Kyushu (Ryukyu and Ogasawara) have significant levels of phylogenetic endemism, with low (non-significant) endemism observed elsewhere. The vast majority of endemic grid cells are mixed, with very few paleo- or neoendemic cells. Small clusters of significantly endemic cells were also observed northwest of Tokyo (Nagano prefecture), and in central Hokkaido (Daisetsuzan mountains). When we ran CANAPE with a reduced dataset including only taxa endemic to Japan to account for the border effect, there were fewer significant cells, but patterns were otherwise similar to the full dataset (`r s_figure("endemism-restricted")`).

### Bioregions

Analysis of biogeographic regions using taxonomy (species names) resulted in the vast majority of grid cells classified into four regions: Cluster 1 (Hokkaido and northern Honshu), Cluster 2 (southern Honshu), Cluster 3 (Okinawa), and Cluster 4 (Ogasawara Islands) (`r figure("bioregions")`A). Using phylogenetic distances produced similar results, but Okinawa and the Ogasawara Islands merged into a single cluster (`r figure("bioregions")`B). The other much smaller clusters are likely artifacts of insufficiently low taxon richness for accurate clustering in a small number of grid cells, and are not discussed further.

Taxonomic clusters 3 and 4 have higher SES values of PD, RPD, FD, and RFD than clusters 1 and 2 (`r figure("div-by-tax-region")`). Clusters 3 and 4 also have consistently high PE *p*-scores, whereas clusters 1 and 2 have a much wider variation. Cluster 2 has lower RPD and RFD than cluster 1. Comparison of biodiversity metrics across phylogenetic clusters was similar, except that phylogenetic cluster 3 mostly corresponds to taxonomic clusters 3 and 4 combined.

### Conservation status

```{r conserv-stats}
# Convert signif_cells_protected_area to wide format
signif_cells_protected_area_wide <-
signif_cells_protected_area %>% select(status, metric, percent_protected) %>%
  pivot_wider(names_from = status, values_from = percent_protected) %>%
  rowwise %>%
  mutate(total = sum(medium, high)) %>%
  ungroup

# Calculate min and max percent area protected overall
min_protected_med <- signif_cells_protected_area_wide %>% pull(medium) %>% min %>% scales::percent(accuracy = 0.1)
max_protected_med <- signif_cells_protected_area_wide %>% pull(medium) %>% max %>% scales::percent(accuracy = 0.1)
min_protected_high <- signif_cells_protected_area_wide %>% pull(high) %>% min %>% scales::percent(accuracy = 0.1)
max_protected_high <- signif_cells_protected_area_wide %>% pull(high) %>% max %>% scales::percent(accuracy = 0.1)
min_protected_total <- signif_cells_protected_area_wide %>% pull(total) %>% min %>% scales::percent(accuracy = 0.1)
max_protected_total <- signif_cells_protected_area_wide %>% pull(total) %>% max %>% scales::percent(accuracy = 0.1)

# Calculate min and max protected area for the best protected category
signif_cells_protected_area_best <-
  signif_cells_protected_area_wide %>%
  slice_max(total)

# Verify that the best-protected biodiv type is PE
assertthat::assert_that(signif_cells_protected_area_best$metric == "pe")

best_protected_med <- signif_cells_protected_area_best %>% pull(medium) %>% scales::percent(accuracy = 0.1)
best_protected_high <- signif_cells_protected_area_best %>% pull(high) %>% scales::percent(accuracy = 0.1)
best_protected_total <- signif_cells_protected_area_best %>% pull(total) %>% scales::percent(accuracy = 0.1)

# Calculate the total area of Japan
area_of_japan <-
  japan_shp %>%
  mutate(area = st_area(.) %>% units::set_units(km^2)) %>%
  pull(area)

# Calculate baseline protection percentage: percent protected areas across the whole country
total_percent_protected <-
  protected_areas %>%
  as_tibble() %>%
  select(-geometry) %>%
  filter(status %in% c("medium", "high")) %>%
  group_by(status) %>%
  summarize(
    total_area = sum(area),
    .groups = "drop"
  ) %>%
  mutate(percent_protected = total_area/area_of_japan) %>%
  mutate(percent_protected = as.numeric(percent_protected))

# Extract values for medium and high protection level
total_percent_protected_medium <- total_percent_protected %>% filter(status == "medium") %>% pull(percent_protected)
total_percent_protected_high <- total_percent_protected %>% filter(status == "high") %>% pull(percent_protected)
total_percent_protected_all <- sum(total_percent_protected_medium, total_percent_protected_high)
```

The percentage of protected area in cells with significantly high biodiversity range from `r min_protected_total` to `r max_protected_total` total, including `r min_protected_high` to `r max_protected_high` with high status and `r min_protected_med` to `r max_protected_med` with medium status (`r figure("conserv-status")`). Areas with high PE were the best protected (`r best_protected_high` high, `r best_protected_med` medium, `r best_protected_total` total). All cells with significantly high biodiversity had a greater percentage of area protected than the overall rate for Japan (`r scales::percent(total_percent_protected_high, accuracy = 0.1)` high status, `r scales::percent(total_percent_protected_medium, accuracy = 0.1)` medium, `r scales::percent(total_percent_protected_all, accuracy = 0.1)` total).

## Discussion

Here, we provide the first insights into the combined taxonomic, phylogenetic, and functional diversity of the ferns of Japan. Our integrative approach reveals distinct patterns of biodiversity in different parts of the archipelago and across biodiversity metrics. These clear differences are reflected in the bioregionalization results, which delimit three to four major bioregions. Finally, we consider the current conservation status of ferns in Japan and the implications of our findings for future conservation activity.

### Distribution of biodiversity

The latitudinal hump-shaped pattern of species richness observed in ferns here has been well-documented [@Kusumoto2017; @Ebihara2019b], and similar patterns have been observed in deciduous broadleaved trees and conifers [@Shiono2015]. This pattern is probably due at least in part to the distribution of land area in Japan: maximum land area is located at mid-latitudes, and area is related to species richness by the well-known species-area relationship [@Arrhenius1921]. There may also be a combined climatic and elevational effect: the southern part of Honshu is warmer and more humid than areas further north, while also harboring high elevational variation that may support a wide range of taxa, which is not present in the southern islands.

The overwhelming pattern characterizing phylogenetic and functional diversity is the high diversity of southern islands (Ryukyu and Bonin Isl.). This is almost certainly due to the presence of primarily tropical lineages that do not occur in other parts of the country; _ genera and _ families only occur south of 30.1° latitude (just south of Yakushima Isl.). The similarity in patterns of PD and FD is likely due to the moderate degree of phylogenetic signal present in the traits used in our analysis, and suggests that in ferns, phylogeny can be used as a reasonable stand-in for functional diversity. Given their high phylogenetic diversity, it is perhaps not surprising that southern islands also host high amounts of phylogenetic endemism; indeed, the vast majority of cells with significant levels of PE are located in the southern islands. This is clearly due to the small land area of these islands, which are the only areas with a subtropical climate in Japan.

Another striking pattern revealed here is the predominance of long branches in the northern half of Honshu and in southern islands, and the predominance of short branches in the southern half on Honshu, particularly Kyushu. The long branches in the southern islands are likely being driven by the tropical lineages mentioned previously; those in the north may be due to lineages that are distantly related to others and become more species rich at high-latitudes, such as *Equisetum* and *Ophioglossum*. The preponderance of short branches in southern Honshu may be due to the radiation of certain species-rich genera such as *Polystichum* and *Cyrtomimum*.

The variation in distribution of different biodiversity metrics is reflected by the bioregions analysis, which tended to group cells with similar biodiversity values. Interestingly, the bioregions identified here correspond roughly to major forest types categorized on an *ad-hoc* basis; Cluster 1 corresponds to the cool or subartic deciduous type, Cluster 2 to the warm evergreen type, and Clusters 3 and 4 to the subtropical rainy type [@Shimizu2014]. This suggests that ferns are useful as possible bioindicators in Japan. Although some studies have emphasized the importance of deep-sea straits (the Tsugaru Strait separating Hokkaido from the rest of Honshu and the Tokara Strait separating the southern islands from Kyushu) in structuring biological communities in Japan [@Millien-Parra1999; @Kubota2014], these did not play a major role in structuring fern bioregions. The Tsugaru Strait had no relation to any of the bioregions, and although the Tokara Strait somewhat splits clusters 2 and 3, there is some overlap between them on the Amami-Oshima Isl. The weak effect of these straits as barriers makes sense given the presumably high dispersal abilities of ferns, which disperse via tiny spores carried on the wind.

### Conservation status and future outlook

Although the different facets of biodiversity have different distribution patterns, we find that grid cells with significantly high amounts of biodiversity tend to be relatively well protected compared to the baseline protection rate, regardless of biodiversity metric. While it is not clear why this is so---it is unlikely that ferns were emphasized as targets during the establishment of conservation zones---one possibility (as suggested previously) is that ferns may serve as bioindicators reflecting overall levels of biodiversity, particularly for other plant groups. Despite the good news that ferns enjoy relatively strong protection status in Japan, we must bear in mind that there are other threats posed that conservation zones cannot ameliorate, the most obvious being climate change. One such other threat is herbivory by native Japanese deer, *Cervus nippon*. Populations of these deer have been increasing exponentially since the 1970s due to a combination of climate change and a decrease in hunting and natural predators. The deer are capable of decimating the herb layer when populations become dense and have been causing extensive damage even within national parks, which have the highest protection status. Although various efforts to prevent deer herbivory have been deployed such as exclusion fences, much work remains to be done to save vulnerable plant populations.

There are a number of caveats that must be kept in mind when interpreting these results. First, the data include species collected over a wide time period, so they do not necessarily represent the current distribution of ferns in Japan. Furthermore, the collections are not evenually collected throughout the country, but rather some areas have many more specimens than others. Indeed, one promising avenue of research may be to take advantage of the most densely sampled areas to investigate changes in abundance over time.

<!-- hybrid taxa, gametophytes, conclusion -->

## Acknowledgements

## Author Contribution

## References
<!-- Insert <div id="refs"></div> to place the refs here instead of at the end -->
<div id="refs"></div>

\clearpage

<!-- Figures -->

```{r plot-prep}
# Change cluster to factor
biodiv_ferns_spatial <- biodiv_ferns_spatial %>%
  mutate(taxonomic_cluster = as.factor(taxonomic_cluster) %>% fct_infreq %>% as.numeric %>% as.factor) %>%
  mutate(phylo_cluster = as.factor(phylo_cluster) %>% fct_infreq %>% as.numeric %>% as.factor) %>%
  # Add redundancy
  mutate(redundancy = 1 - (richness/abundance))

# Download high-res map of Japan,
# crop to spatial div results area
japan <- sf::st_crop(japan_shp, sf::st_bbox(biodiv_ferns_spatial))

# Set CRS
japan_crs <- sf::st_crs(japan)

biodiv_ferns_spatial <- sf::st_set_crs(biodiv_ferns_spatial, japan_crs)

# Set color palettes
# Randomization test significance
signif_cols <-
  c(
    "> 0.99" = "#e31a1c", 
    "> 0.975" = "#fb9a99",
    "< 0.01" = "#1f78b4", 
    "< 0.025" = "#a6cee3",
    "not significant" = "#ffffcc"
  )

# CANAPE
canape_cols <-
  c(
    "neo" = "#e31a1c", 
    "paleo" = "#1f78b4",
    "mixed" = "#6a51a3",
    "super" = "#4a1486",
    "not significant" = "#ffffcc"
  )

# Bioregions
bioregion_cols <- rcartocolor::carto_pal(9, "Bold")[1:8] %>%
  set_names(1:8)
```

```{r raw-diversity, width = 7, height = 4}
# Richness
a <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = richness), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**A** Richness</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "No. taxa") +
  scale_fill_scico(palette = "lajolla") + 
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  map_theme()

# Raw PD
b <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = pd_obs), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**B** PD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "Relative\nbranch\nlength") +
  scale_fill_scico(palette = "lajolla") + 
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  map_theme() +
  theme(
    axis.text.y = element_blank()
  )

# Raw FD
c <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = fd_obs), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**C** FD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "Relative\nbranch\nlength") +
  scale_fill_scico(palette = "lajolla") + 
  scale_x_continuous(breaks = c(125, 135, 145)) +
  scale_y_continuous(breaks = c(25, 35, 45)) +
  map_theme() +
  theme(
    axis.text.y = element_blank()
  )

a + b + c +
  plot_layout(nrow = 1)

ggsave(result_file(figure("raw-diversity"), "pdf"), width = 7, height = 4, units = "in")
```

**`r figure("raw-diversity")`**. `r figure_cap("raw-diversity")` For **B** and **C**, raw branchlengths were transformed to relative values by dividing each branchlength by the sum of all branchlengths.

\clearpage

```{r rand-diversity, fig.width = 7, fig.height = 9}
# PD
a <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = pd_signif), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**A** PD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  scale_fill_manual(values = signif_cols) +
  map_theme() +
  theme(
    axis.text.x = element_blank(),
    legend.title = element_blank()
    )

# RPD
b <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = rpd_signif), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**B** RPD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  map_theme() +
  theme(
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    legend.title = element_blank()
  )

# FD
c <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = fd_signif), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**C** FD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
  map_theme() +
  theme(legend.title = element_blank())

# RFD
d <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = rfd_signif), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**D** RFD</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "") +
  scale_fill_manual(values = signif_cols) +
  scale_x_continuous(guide = guide_axis(check.overlap = TRUE)) +
  map_theme() +
  theme(
    axis.text.y = element_blank(),
    legend.title = element_blank()
  )

a + b + c + d + 
  plot_layout(ncol = 2, nrow = 2, guides = 'collect') &
  theme(legend.position = "bottom")
  
ggsave(result_file(figure("rand-diversity"), "pdf"), width = 7, height = 8, units = "in")
```

**`r figure("rand-diversity")`**. `r figure_cap("rand-diversity")` **A**, Phylogenetic diversity (PD). **B**, Relative phylogenetic diveristy (RPD). **C**, Functional diveristy (FD). **D**, relative functional diveristy (RFD). For each metric, raw values were compared to a null distribution of 999 random values, and the *P*-value calculated using a two-tailed test. For details, see Methods.

\clearpage

```{r endemism, fig.width = 5, fig.height = 5}
ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = endem_type), color = "transparent", alpha = 0.8) +
  scale_fill_manual(values = canape_cols) +
  map_theme() +
  theme(legend.title = element_blank())

ggsave(result_file(figure("endemism"), "pdf"), width = 5, height = 5, units = "in")
```

**`r figure("endemism")`**. `r figure_cap("endemism")` 'paleo' indicates paleoendemic areas (those with an overabundance of long branches); 'neo' indicates neodendemic areas (those with an overabundance of short branches); 'mixed' indicates significantly endemic sites that have neither an overabundance of short nor long branches; 'super' indicates highly significantly endemic sites that have neither an overabundance of short nor long branches. For details, see Methods.

\clearpage

```{r bioregions, fig.width = 7, fig.height = 5}
# Taxonomic bioregions
a <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = taxonomic_cluster), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**A** Taxonomic</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "Cluster") +
  scale_fill_manual(values = bioregion_cols) +
  map_theme()

# Phylogenetic bioregions
b <- ggplot() +
  geom_sf(data = japan, fill = "grey90", color = "black", lwd = 0.25) +
  geom_sf(data = biodiv_ferns_spatial, aes(fill = phylo_cluster), color = "transparent", alpha = 0.8) +
  annotate(geom="richtext", y = Inf, x = -Inf, label = "<span style='color:black'>**B** Phylogenetic</span>", hjust = "inward", vjust = "inward", color = "white") + 
  labs(fill = "Cluster") +
  scale_fill_manual(values = bioregion_cols) +
  map_theme() +
  theme(
    axis.text.y = element_blank()
  )

a + b +
  plot_layout(ncol = 2)

ggsave(result_file(figure("bioregions"), "pdf"), width = 7, height = 5, units = "in")
```

**`r figure("bioregions")`**. `r figure_cap("bioregions")` **A**, taxonomic bioregions. **B**, phylogenetic bioregions. Bioregions determined by clustering taxonomic (Sørensen) or phylogenetic (PhyloSor) distances into *k* = 8 or 7 groups, respectively, using the R package \package{phyloregion}.

\clearpage

```{r div-by-tax-region, fig.width = 7, fig.height = 7}
data <-
biodiv_ferns_spatial %>%
  filter(taxonomic_cluster %in% c(1,2,3,4)) %>%
  as_tibble %>%
  select(pd_obs_z, rpd_obs_z, fd_obs_z, rfd_obs_z, pe_obs_p_upper, taxonomic_cluster)

cluster_cols <- carto_pal(8, "Bold")[1:4] %>%
  set_names(1:4)

a <- ggplot(data, aes(x = taxonomic_cluster, y = pd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES PD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

b <- ggplot(data, aes(x = taxonomic_cluster, y = rpd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RPD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

c <- ggplot(data, aes(x = taxonomic_cluster, y = fd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES FD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

d <- ggplot(data, aes(x = taxonomic_cluster, y = rfd_obs_z)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "SES RFD",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.title.x = element_blank()
  )

e <- ggplot(data, aes(x = taxonomic_cluster, y = pe_obs_p_upper)) +
  geom_jitter(aes(color = taxonomic_cluster)) +
  scale_color_manual(values = cluster_cols, breaks = 1:4) +
  geom_boxplot(color = "grey80", fill = "transparent", outlier.shape = NA, alpha = 0.25) +
  labs(
    y = "PE p-score",
    x = "Cluster",
    color = "Cluster"
  ) +
  theme_bw() +
  theme(
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.y = element_blank(),
  )

a + b + c + d + e +
  plot_layout(ncol = 2) +
plot_annotation(
  tag_levels = 'A'
) &
  theme(legend.position = "none")
  
ggsave(result_file(figure("div-by-tax-region"), "pdf"), width = 7, height = 7, units = "in")

```

**`r figure("div-by-tax-region")`**. `r figure_cap("div-by-tax-region")` **A**, Standard effect size (SES) of phylogenetic diversity (PD). **B**, SES of relative PD (RPD). **C**, SES of functional diversity (FD). **D**, SES of relative FD (RFD). **E**, p-score of observed phylogenetic endemism (PE) relative to 999 random communities. Only top four clusters with the most sites shown.

\clearpage

```{r conserv-status, fig.width = 4, fig.height = 4}
# Plot protected areas
signif_cells_protected_area %>%
  mutate(
    metric = case_when(
      metric == "pd" ~ "PD",
      metric == "fd" ~ "FD",
      metric == "pe" ~ "PE",
      metric == "richness" ~ "Richness",
    ),
    status = str_to_sentence(status)) %>%
  mutate(metric = factor(metric, levels = c("PE", "FD", "PD", "Richness"))) %>%
  ggplot(aes(x = percent_protected, y = metric, fill = status)) +
  geom_col(position = position_dodge()) +
  geom_vline(xintercept = total_percent_protected_medium, color = "grey40", linetype = 3) +
  geom_vline(xintercept = total_percent_protected_high, color = "grey40", linetype = 2) +
  scale_fill_viridis_d() +
  scale_x_continuous(labels = scales::percent) +
  labs(
    x = "Percent protected",
    y = "Biodiversity metric",
    fill = "Protection status"
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "bottom"
  )

ggsave(result_file(figure("conserv-status"), "pdf"), width = 4, height = 4, units = "in")

```

**`r figure("conserv-status")`**. `r figure_cap("conserv-status")` Biodiversity metrics include phylogenetic diversity (PD), functional diversity (FD), relative PD (RPD), relative FD (RFD), and phylogenetic endemism (PE). Significance of biodiversity metrics assessed by a one-tailed test comparing observed values to a null distribution of 999 random values (see Methods). Vertical lines indicate total percent area protected across Japan regardless of degree of biodiversity; dashed, high protection level (`r scales::percent(total_percent_protected_high, accuracy = 0.1)`); dotted, medium protection level (`r scales::percent(total_percent_protected_medium, accuracy = 0.1)`).
